diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..3093598
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,3 @@
+*.swp
+.vs/
+build/
diff --git a/.travis.yml b/.travis.yml
index 26b5a00..8407274 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -1,44 +1,35 @@
-\language: cpp
-env:
-  global: 
-    - secure: "WEPiGQTAVh02Fc5Qxvj2VBZoWBgNjjB6w5MdNwg6NtzFKhbhxB1OqfaVE7EJV1Cw9mf1k5Qcg2fSy9bQEJllWNclRpbRGfEcC+Crx3CyjFOebSr6e6T7lAjIspIVSRf9+y6oPEY2zR8yXoX/HAEYyw3IQzJtKCHIbSJrrZsPomc="
-    
+sudo: required
+language: cpp
 matrix:
   include:
     - os: linux
+      dist: trusty
+      compiler: clang
+    - os: linux
+      dist: trusty
       addons:
         apt:
+          sources:
+            - ubuntu-toolchain-r-test
           packages:
-            - gfortran
-    - os: osx 
-      osx_image: xcode8.2
-      env: OSX=10.12
-      compiler: clang
+            - g++-6
+      compiler: g++-6
     - os: osx
-      osx_image: xcode8
-      env: OSX=10.11
       compiler: clang
     - os: osx
-      osx_image: xcode6.4
-      env: OSX=10.10
-      compiler: clang
-  allow_failures:
-    - os: osx
+      compiler: gcc
 
-before_script:
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then  export PLATFORM=osx$OSX-x86_64-clang`clang -dumpversion`; fi
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update; brew install bash gcc; brew link --overwrite gcc; gfortran --version; fi
-  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then export PLATFORM=linux-x86_64-gcc`gcc -dumpversion`; fi
-  - git clone https://github.com/coin-or-tools/BuildTools
-  - bash BuildTools/get.dependencies.sh fetch > /dev/null
+before_install:
+  - eval "${MATRIX_EVAL}"
+  - if [ $TRAVIS_OS_NAME == linux ]; then sudo apt-get -qq update; fi
+  - if [ $TRAVIS_OS_NAME == linux ]; then sudo apt-get -y install zlib1g-dev; fi
 
-script:
-  - bash BuildTools/get.dependencies.sh build --verbosity=2 --test
+  - if [ $TRAVIS_OS_NAME == osx ]; then sudo chown -R $(whoami) /usr/local; fi
+  # Avoid bug on CI with OSX, see https://github.com/travis-ci/travis-ci/issues/8552
+  - if [ $TRAVIS_OS_NAME == osx ]; then brew update; fi
+  - if [ $TRAVIS_OS_NAME == osx ]; then brew install zlib; fi
 
-after_script:
-  - if [ $TRAVIS_BRANCH = "master" ]; then export VERSION=trunk; else export VERSION=`echo $TRAVIS_BRANCH | cut -d "/" -f 2`; fi
-  - export PROJECT=`echo $TRAVIS_REPO_SLUG | cut -d "/" -f 2`
-  - export TGZ_FILE=$PROJECT-$VERSION-$PLATFORM.tgz
-  - echo $TGZ_FILE
-  - tar -czvf $TGZ_FILE build/lib/* build/bin/* build/include/* build/share/* README.md INSTALL LICENSE Cbc/AUTHORS
-  - curl -T $TGZ_FILE -utkralphs:$BINTRAY_API -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/coin-or/download/$PROJECT/$VERSION/$TGZ_FILE
\ No newline at end of file
+script:
+  - cmake --version
+  - cmake -H. -Bbuild -DCMAKE_VERBOSE_MAKEFILE=ON
+  - cmake --build build --target all
diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..7663d60
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,197 @@
+cmake_minimum_required(VERSION 3.0.2)
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE "Release" CACHE STRING
+		"Choose the type of build, options are: Debug Release RelWithDebInfo MinSizeRel."
+		FORCE)
+endif(NOT CMAKE_BUILD_TYPE)
+
+# Not a subproject -> add find_package macro
+if ("^${CMAKE_SOURCE_DIR}$" STREQUAL "^${CMAKE_CURRENT_SOURCE_DIR}$")
+	# Use find_package everywhere, no-op if it's a target
+	macro(find_package)
+		if(NOT TARGET ${ARGV0} AND NOT TARGET ${ARGV0}::${ARGV0})
+			_find_package(${ARGV})
+		else()
+			if(TARGET ${ARGV0})
+				get_target_property(TGT_VER ${ARGV0} VERSION)
+				set(TGT ${ARGV0})
+			else()
+				get_target_property(TGT_VER ${ARGV0}::${ARGV0} VERSION)
+				set(TGT ${ARGV0}::${ARGV0})
+			endif()
+			message(STATUS "Found ${ARGV0}: CMake Target ${TGT} (found version \"${TGT_VER}\")")
+			set(${ARGV0}_FOUND TRUE)
+		endif()
+	endmacro()
+endif()
+
+project(Coin VERSION 2.9.9 LANGUAGES C CXX)
+message(STATUS "Project: Coin")
+
+include(CTest)
+include(ParseAc)
+
+# ZLIB
+find_package(ZLIB)
+if(ZLIB_FOUND)
+	set(HAVE_ZLIB_H "1" CACHE INTERNAL "Use zlib")
+	set(COIN_HAS_ZLIB "1" CACHE INTERNAL "Use zlib")
+endif()
+
+# Flags
+include (CheckIncludeFile)
+check_include_file(math.h HAVE_MATH_H)
+check_include_file(ctype.h HAVE_CTYPE_H)
+check_include_file(inttypes.h HAVE_INTTYPES_H)
+check_include_file(float.h HAVE_FLOAT_H)
+check_include_file(ieeefp.h HAVE_IEEEFP_H)
+check_include_file(stdarg.h HAVE_STDARG_H)
+check_include_file(stddef.h HAVE_STDDEF_H)
+check_include_file(stdint.h HAVE_STDINT_H)
+check_include_file(stdio.h HAVE_STDIO_H)
+check_include_file(stdlib.h HAVE_STDLIB_H)
+check_include_file(assert.h HAVE_ASSERT_H)
+check_include_file(dlfcn.h HAVE_DLFCN_H)
+check_include_file(endian.h HAVE_ENDIAN_H)
+check_include_file(memory.h HAVE_MEMORY_H)
+check_include_file(strings.h HAVE_STRINGS_H)
+check_include_file(string.h HAVE_STRING_H)
+check_include_file(time.h HAVE_TIME_H)
+check_include_file(unistd.h HAVE_UNISTD_H)
+check_include_file(sys/stat.h HAVE_SYS_STAT_H)
+check_include_file(sys/types.h HAVE_SYS_TYPES_H)
+
+include (CheckIncludeFileCXX)
+check_include_file_cxx(cmath HAVE_CMATH)
+check_include_file_cxx(cctype HAVE_CCTYPE)
+check_include_file_cxx(cinttypes HAVE_CINTTYPES)
+check_include_file_cxx(cfloat HAVE_CFLOAT)
+check_include_file_cxx(cieeefp HAVE_CIEEEFP)
+check_include_file_cxx(cstdarg HAVE_CSTDARG)
+check_include_file_cxx(cstddef HAVE_CSTDDEF)
+check_include_file_cxx(cstdint HAVE_CSTDINT)
+check_include_file_cxx(cstdio HAVE_CSTDIO)
+check_include_file_cxx(cstdlib HAVE_CSTDLIB)
+check_include_file_cxx(cassert HAVE_CASSERT)
+check_include_file_cxx(cstring HAVE_CSTRING)
+check_include_file_cxx(ctime HAVE_CTIME)
+
+set(STDC_HEADERS 1 CACHE INTERNAL "System has ANSI C header files")
+
+set(TEST_INCLUDES "")
+if(HAVE_CMATH)
+	list(APPEND TEST_INCLUDES "cmath")
+endif()
+if(HAVE_CFLOAT)
+	list(APPEND TEST_INCLUDES "cfloat")
+endif()
+if(HAVE_CIEEEFP)
+	list(APPEND TEST_INCLUDES "cieeefp")
+endif()
+if(HAVE_MATH_H)
+	list(APPEND TEST_INCLUDES "math.h")
+endif()
+if(HAVE_FLOAT_H)
+	list(APPEND TEST_INCLUDES "float.h")
+endif()
+if(HAVE_IEEEFP_H)
+	list(APPEND TEST_INCLUDES "ieeefp.h")
+endif()
+
+# ISFINITE
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles(
+  "#include <cmath>\nint main(){return std::isfinite(0);}"
+  HAVE_STD_ISFINITE)
+include(CheckFunctionExists)
+include(CheckCXXSymbolExists)
+check_cxx_symbol_exists(isfinite "${TEST_INCLUDES}" HAVE_ISFINITE)
+check_cxx_symbol_exists(finite   "${TEST_INCLUDES}" HAVE_FINITE)
+check_cxx_symbol_exists(_finite  "${TEST_INCLUDES}" HAVE__FINITE)
+check_cxx_symbol_exists(__finite "${TEST_INCLUDES}" HAVE___FINITE)
+if(HAVE_STD_ISFINITE)
+	set(COIN_C_FINITE "std::isfinite")
+elseif(HAVE_ISFINITE)
+	set(COIN_C_FINITE "isfinite")
+elseif(HAVE_FINITE)
+	set(COIN_C_FINITE "finite")
+elseif(HAVE__FINITE)
+	set(COIN_C_FINITE "_finite")
+elseif(HAVE___FINITE)
+	set(COIN_C_FINITE "__finite")
+else()
+	message(FATAL_ERROR "Can't find isfinite()")
+endif()
+message(STATUS "Found isfinite: ${COIN_C_FINITE}")
+
+# ISNAN
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles(
+  "#include <cmath>\nint main(){return std::isnan(0);}"
+  HAVE_STD_ISNAN)
+include(CheckFunctionExists)
+include(CheckCXXSymbolExists)
+check_cxx_symbol_exists(isnan   "${TEST_INCLUDES}" HAVE_ISNAN)
+check_cxx_symbol_exists(_isnan  "${TEST_INCLUDES}" HAVE__ISNAN)
+check_cxx_symbol_exists(__isnan "${TEST_INCLUDES}" HAVE___ISNAN)
+if(HAVE_STD_ISNAN)
+	set(COIN_C_ISNAN "std::isnan")
+elseif(HAVE_ISNAN)
+	set(COIN_C_ISNAN "isnan")
+elseif(HAVE__ISNAN)
+	set(COIN_C_ISNAN "_isnan")
+elseif(HAVE___ISNAN)
+	set(COIN_C_ISNAN "__isnan")
+else()
+	message(FATAL_ERROR "Can't find isnan()")
+endif()
+message(STATUS "Found isnan: ${COIN_C_ISNAN}")
+
+# Basic type
+include(CheckTypeSize)
+check_type_size("int64_t" SIZEOF_INT64_T)
+check_type_size("long long" SIZEOF_LONG_LONG)
+check_type_size("long" SIZEOF_LONG)
+check_type_size("uint64_t" SIZEOF_UINT64_T)
+check_type_size("unsigned long long" SIZEOF_ULONG_LONG)
+check_type_size("unsigned long" SIZEOF_ULONG)
+check_type_size("intptr_t" SIZEOF_INTPTR_T)
+check_type_size("int *" SIZEOF_INT_P)
+
+if(SIZEOF_INT64_T EQUAL "8")
+  set(COIN_INT64_T "int64_t")
+elseif(SIZEOF_LONG EQUAL "8")
+  set(COIN_INT64_T "long")
+elseif(SIZEOF_LONG_LONG EQUAL "8")
+  set(COIN_INT64_T "long long")
+else()
+	message(FATAL_ERROR "Can't find suitable int64_t")
+endif()
+message(STATUS "Found int64_t: ${COIN_INT64_T}")
+
+if(SIZEOF_UINT64_T EQUAL "8")
+  set(COIN_UINT64_T "uint64_t")
+elseif(SIZEOF_ULONG EQUAL "8")
+  set(COIN_INT64_T "unsigned long")
+elseif(SIZEOF_ULONG_LONG EQUAL "8")
+  set(COIN_INT64_T "unsigned long long")
+else()
+	message(FATAL_ERROR "Can't find suitable uint64_t")
+endif()
+message(STATUS "Found uint64_t: ${COIN_UINT64_T}")
+
+if(SIZEOF_INTPTR_T)
+	set(COIN_INTPTR_T "intptr_t")
+elseif(SIZEOF_INT_P)
+  set(COIN_INTPTR_T "int *")
+else()
+	message(FATAL_ERROR "Can't find suitable intptr_t")
+endif()
+message(STATUS "Found intptr_t: ${COIN_INTPTR_T}")
+
+add_subdirectory(CoinUtils)
+add_subdirectory(Osi)
+add_subdirectory(Clp)
+add_subdirectory(Cgl)
+add_subdirectory(Cbc)
diff --git a/Cbc/CMakeLists.txt b/Cbc/CMakeLists.txt
new file mode 100644
index 0000000..bb857fa
--- /dev/null
+++ b/Cbc/CMakeLists.txt
@@ -0,0 +1,275 @@
+message(STATUS "Subproject: Cbc")
+
+find_package(CoinUtils REQUIRED)
+find_package(Osi REQUIRED)
+find_package(Clp REQUIRED)
+find_package(Cgl REQUIRED)
+
+set(NAME "CBC")
+parse_ac(VERSION MAJOR MINOR PATCH)
+
+set(COIN_${NAME}_CHECKLEVEL "0" CACHE INTERNAL
+	"${NAME} check level")
+set(COIN_${NAME}_VERBOSITY "0" CACHE INTERNAL
+	"${NAME} verbosity level")
+configure_file(config.h.cmake.in config.h)
+configure_file(config_cbc.h.cmake.in config_cbc.h)
+
+set(_SRCS
+	#src/CbcBranchBase.cpp
+	src/CbcBranchAllDifferent.cpp
+	src/CbcBranchCut.cpp
+	src/CbcBranchDecision.cpp
+	src/CbcBranchDefaultDecision.cpp
+	src/CbcBranchDynamic.cpp
+	src/CbcBranchingObject.cpp
+	src/CbcBranchLotsize.cpp
+	src/CbcBranchToFixLots.cpp
+	src/CbcClique.cpp
+	src/CbcCompareDefault.cpp
+	src/CbcCompareDepth.cpp
+	src/CbcCompareEstimate.cpp
+	src/CbcCompareObjective.cpp
+	src/CbcConsequence.cpp
+	src/CbcCountRowCut.cpp
+	src/CbcCutGenerator.cpp
+	src/CbcCutModifier.cpp
+	src/CbcCutSubsetModifier.cpp
+	src/CbcDummyBranchingObject.cpp
+	src/CbcEventHandler.cpp
+	src/CbcFathom.cpp
+	src/CbcFathomDynamicProgramming.cpp
+	src/CbcFixVariable.cpp
+	src/CbcFollowOn.cpp
+	src/CbcFullNodeInfo.cpp
+	src/CbcGenBaB.cpp
+	src/CbcGenCbcParam.cpp
+	src/CbcGenCbcParamUtils.cpp
+	src/CbcGenCtlBlk.cpp
+	src/CbcGeneral.cpp
+	src/CbcGeneralDepth.cpp
+	src/CbcGenMessages.cpp
+	src/CbcGenOsiParam.cpp
+	src/CbcGenOsiParamUtils.cpp
+	src/CbcGenParam.cpp
+	src/CbcGenParamUtils.cpp
+	src/CbcGenSolution.cpp
+	src/CbcGenSolvers.cpp
+	src/CbcHeuristic.cpp
+	src/CbcHeuristicDINS.cpp
+	src/CbcHeuristicDiveCoefficient.cpp
+	src/CbcHeuristicDive.cpp
+	src/CbcHeuristicDiveFractional.cpp
+	src/CbcHeuristicDiveGuided.cpp
+	src/CbcHeuristicDiveLineSearch.cpp
+	src/CbcHeuristicDivePseudoCost.cpp
+	src/CbcHeuristicDiveVectorLength.cpp
+	src/CbcHeuristicDW.cpp
+	src/CbcHeuristicFPump.cpp
+	src/CbcHeuristicGreedy.cpp
+	src/CbcHeuristicLocal.cpp
+	src/CbcHeuristicPivotAndFix.cpp
+	src/CbcHeuristicRandRound.cpp
+	src/CbcHeuristicRENS.cpp
+	src/CbcHeuristicRINS.cpp
+	src/CbcHeuristicVND.cpp
+	src/CbcMessage.cpp
+	src/CbcModel.cpp
+	src/CbcNode.cpp
+	src/CbcNodeInfo.cpp
+	src/CbcNWay.cpp
+	src/CbcObject.cpp
+	src/CbcObjectUpdateData.cpp
+	src/CbcParam.cpp
+	src/CbcPartialNodeInfo.cpp
+	src/CbcSimpleInteger.cpp
+	src/CbcSimpleIntegerDynamicPseudoCost.cpp
+	src/CbcSimpleIntegerPseudoCost.cpp
+	src/CbcSOS.cpp
+	src/CbcStatistics.cpp
+	src/CbcStrategy.cpp
+	src/CbcSubProblem.cpp
+	src/CbcThread.cpp
+	src/CbcTree.cpp
+	src/CbcTreeLocal.cpp)
+
+set(_HDRS
+	src/CbcConfig.h
+	src/CbcBranchActual.hpp
+	src/CbcBranchAllDifferent.hpp
+	src/CbcBranchBase.hpp
+	src/CbcBranchDynamic.hpp
+	src/CbcBranchCut.hpp
+	src/CbcBranchDecision.hpp
+	src/CbcBranchDefaultDecision.hpp
+	src/CbcBranchingObject.hpp
+	src/CbcBranchLotsize.hpp
+	src/CbcBranchToFixLots.hpp
+	src/CbcCompareActual.hpp
+	src/CbcCompareBase.hpp
+	src/CbcCompareDefault.hpp
+	src/CbcCompareDepth.hpp
+	src/CbcCompareEstimate.hpp
+	src/CbcCompareObjective.hpp
+	src/CbcConsequence.hpp
+	src/CbcClique.hpp
+	src/CbcCompare.hpp
+	src/CbcCountRowCut.hpp
+	src/CbcCutGenerator.hpp
+	src/CbcCutModifier.hpp
+	src/CbcCutSubsetModifier.hpp
+	src/CbcDummyBranchingObject.hpp
+	src/CbcFathom.hpp
+	src/CbcEventHandler.hpp
+	src/CbcFathomDynamicProgramming.hpp
+	src/CbcFeasibilityBase.hpp
+	src/CbcFixVariable.hpp
+	src/CbcFollowOn.hpp
+	src/CbcFullNodeInfo.hpp
+	src/CbcGeneral.hpp
+	src/CbcGeneralDepth.hpp
+	src/CbcHeuristic.hpp
+	src/CbcHeuristicDINS.hpp
+	src/CbcHeuristicDive.hpp
+	src/CbcHeuristicDiveCoefficient.hpp
+	src/CbcHeuristicDiveFractional.hpp
+	src/CbcHeuristicDiveGuided.hpp
+	src/CbcHeuristicDiveLineSearch.hpp
+	src/CbcHeuristicDivePseudoCost.hpp
+	src/CbcHeuristicDiveVectorLength.hpp
+	src/CbcHeuristicDW.hpp
+	src/CbcHeuristicFPump.hpp
+	src/CbcHeuristicGreedy.hpp
+	src/CbcHeuristicLocal.hpp
+	src/CbcHeuristicPivotAndFix.hpp
+	src/CbcHeuristicRandRound.hpp
+	src/CbcHeuristicRENS.hpp
+	src/CbcHeuristicRINS.hpp
+	src/CbcHeuristicVND.hpp
+	src/CbcMessage.hpp
+	src/CbcModel.hpp
+	src/CbcNode.hpp
+	src/CbcNodeInfo.hpp
+	src/CbcNWay.hpp
+	src/CbcObject.hpp
+	src/CbcObjectUpdateData.hpp
+	src/CbcPartialNodeInfo.hpp
+	src/CbcSimpleInteger.hpp
+	src/CbcSimpleIntegerDynamicPseudoCost.hpp
+	src/CbcSimpleIntegerPseudoCost.hpp
+	src/CbcStrategy.hpp
+	src/CbcSOS.hpp
+	src/CbcSubProblem.hpp
+	src/CbcTree.hpp
+	src/CbcTreeLocal.hpp
+	src/ClpConstraintAmpl.hpp
+	src/ClpAmplObjective.hpp)
+
+add_library(Cbc ${_SRCS} ${_HDRS})
+target_include_directories(Cbc PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
+target_compile_definitions(Cbc
+	PUBLIC HAVE_CONFIG_H
+	PRIVATE CBC_BUILD)
+if(WIN32)
+	target_compile_definitions(Cbc PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(Cbc PUBLIC cxx_std_11)
+target_link_libraries(Cbc PUBLIC
+	CoinUtils::CoinUtils
+	Osi::Osi
+	Clp::Clp
+	Cgl::Cgl)
+set_target_properties(Cbc PROPERTIES
+	PUBLIC_HEADER "${_HDRS};${CMAKE_CURRENT_BINARY_DIR}/config_cbc.h"
+	VERSION ${VERSION})
+add_library(Cbc::Cbc ALIAS Cbc)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS Cbc
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
+
+##############
+##  OsiCbc  ##
+##############
+set(_SRCS src/OsiCbc/OsiCbcSolverInterface.cpp)
+set(_HDRS	src/OsiCbc/OsiCbcSolverInterface.hpp)
+
+add_library(OsiCbc ${_SRCS} ${_HDRS})
+target_include_directories(OsiCbc PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/OsiCbc>)
+#target_compile_definitions(OsiCbc PRIVATE ...)
+if(WIN32)
+	target_compile_definitions(OsiCbc PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(OsiCbc PUBLIC cxx_std_11)
+target_link_libraries(OsiCbc PUBLIC
+	CoinUtils::CoinUtils
+	Osi::Osi
+	Cbc::Cbc)
+set_target_properties(OsiCbc PROPERTIES
+	PUBLIC_HEADER "${_HDRS}"
+	VERSION ${VERSION})
+add_library(OsiCbc::OsiCbc ALIAS OsiCbc)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS OsiCbc
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
+
+#################
+##  CbcSolver  ##
+#################
+set(_SRCS
+	src/Cbc_C_Interface.cpp
+	src/CbcCbcParam.cpp
+	src/Cbc_ampl.cpp
+	src/CbcLinked.cpp
+	src/CbcLinkedUtils.cpp
+	src/CbcSolverHeuristics.cpp
+	src/CbcSolverAnalyze.cpp
+	src/CbcSolver.cpp
+	src/CbcMipStartIO.cpp
+	src/CbcSolverExpandKnapsack.cpp)
+
+set(_HDRS
+	src/Cbc_C_Interface.h
+	src/Cbc_ampl.h
+	src/CbcLinked.hpp
+	src/CbcSolver.hpp
+	src/CbcMipStartIO.hpp)
+
+add_library(CbcSolver ${_SRCS} ${_HDRS})
+target_include_directories(CbcSolver PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>)
+target_compile_definitions(CbcSolver
+	PRIVATE CBC_THREAD_SAFE CBC_NO_INTERRUPT)
+if(WIN32)
+	target_compile_definitions(CbcSolver PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(CbcSolver PUBLIC cxx_std_11)
+target_link_libraries(CbcSolver PUBLIC
+	Cbc::Cbc)
+set_target_properties(CbcSolver PROPERTIES
+	PUBLIC_HEADER "${_HDRS}"
+	VERSION ${VERSION})
+add_library(CbcSolver::CbcSolver ALIAS CbcSolver)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS CbcSolver
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
diff --git a/Cbc/config.h.cmake.in b/Cbc/config.h.cmake.in
new file mode 100644
index 0000000..719ed8d
--- /dev/null
+++ b/Cbc/config.h.cmake.in
@@ -0,0 +1,115 @@
+/*config.h. Generated by configure_file.*/
+
+#define COIN_HAS_COINUTILS 1
+#define COIN_HAS_OSI 1
+#define COIN_HAS_CLP 1
+#define ${NAME}_DEFAULT_SOLVER "clp"
+
+/* VERSION */
+#define VERSION "${VERSION}"
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+/* HAVE_MATH_H */
+#cmakedefine HAVE_MATH_H ${HAVE_MATH_H}
+/* HAVE_CTYPE_H */
+#cmakedefine HAVE_CTYPE_H ${HAVE_CTYPE_H}
+/* HAVE_INTTYPES_H */
+#cmakedefine HAVE_INTTYPES_H ${HAVE_INTTYPES_H}
+/* HAVE_FLOAT_H */
+#cmakedefine HAVE_FLOAT_H ${HAVE_FLOAT_H}
+/* HAVE_IEEEFP_H */
+#cmakedefine HAVE_IEEEFP_H ${HAVE_IEEEFP_H}
+/* HAVE_STDARG_H */
+#cmakedefine HAVE_STDARG_H ${HAVE_STDARG_H}
+/* HAVE_STDDEF_H */
+#cmakedefine HAVE_STDDEF_H ${HAVE_STDDEF_H}
+/* HAVE_STDINT_H */
+#cmakedefine HAVE_STDINT_H ${HAVE_STDINT_H}
+/* HAVE_STDIO_H */
+#cmakedefine HAVE_STDIO_H ${HAVE_STDIO_H}
+/* HAVE_STDLIB_H */
+#cmakedefine HAVE_STDLIB_H ${HAVE_STDLIB_H}
+/* HAVE_ASSERT_H */
+#cmakedefine HAVE_ASSERT_H ${HAVE_ASSERT_H}
+/* HAVE_DLFCN_H */
+#cmakedefine HAVE_DLFCN_H ${HAVE_DLFCN_H}
+/* HAVE_ENDIAN_H */
+#cmakedefine HAVE_ENDIAN_H ${HAVE_ENDIAN_H}
+/* HAVE_MEMORY_H */
+#cmakedefine HAVE_MEMORY_H ${HAVE_MEMORY_H}
+/* HAVE_STRINGS_H */
+#cmakedefine HAVE_STRINGS_H ${HAVE_STRINGS_H}
+/* HAVE_STRING_H */
+#cmakedefine HAVE_STRING_H ${HAVE_STRING_H}
+/* HAVE_TIME_H */
+#cmakedefine HAVE_TIME_H ${HAVE_TIME_H}
+/* HAVE_UNISTD_H */
+#cmakedefine HAVE_UNISTD_H ${HAVE_UNISTD_H}
+/* HAVE_SYS_STAT_H */
+#cmakedefine HAVE_SYS_STAT_H ${HAVE_SYS_STAT_H}
+/* HAVE_SYS_TYPES_H */
+#cmakedefine HAVE_SYS_TYPES_H ${HAVE_SYS_TYPES_H}
+
+/* HAVE_CMATH */
+#cmakedefine HAVE_CMATH ${HAVE_CMATH}
+/* HAVE_CCTYPE */
+#cmakedefine HAVE_CCTYPE ${HAVE_CCTYPE}
+/* HAVE_CINTTYPES */
+#cmakedefine HAVE_CINTTYPES ${HAVE_CINTTYPES}
+/* HAVE_CFLOAT */
+#cmakedefine HAVE_CFLOAT ${HAVE_CFLOAT}
+/* HAVE_CIEEEFP */
+#cmakedefine HAVE_CIEEEFP ${HAVE_CIEEEFP}
+/* HAVE_CSTDARG */
+#cmakedefine HAVE_CSTDARG ${HAVE_CSTDARG}
+/* HAVE_CSTDDEF */
+#cmakedefine HAVE_CSTDDEF ${HAVE_CSTDDEF}
+/* HAVE_CSTDINT */
+#cmakedefine HAVE_CSTDINT ${HAVE_CSTDINT}
+/* HAVE_CSTDIO */
+#cmakedefine HAVE_CSTDIO ${HAVE_CSTDIO}
+/* HAVE_CSTDLIB */
+#cmakedefine HAVE_CSTDLIB ${HAVE_CSTDLIB}
+/* HAVE_CASSERT */
+#cmakedefine HAVE_CASSERT ${HAVE_CASSERT}
+/* HAVE_CSTRING */
+#cmakedefine HAVE_CSTRING ${HAVE_CSTRING}
+/* HAVE_CTIME */
+#cmakedefine HAVE_CTIME ${HAVE_CTIME}
+
+/* COIN_C_FINITE */
+#cmakedefine COIN_C_FINITE ${COIN_C_FINITE}
+/* COIN_C_ISNAN */
+#cmakedefine COIN_C_ISNAN ${COIN_C_ISNAN}
+/* COIN_INT64_T */
+#cmakedefine COIN_INT64_T ${COIN_INT64_T}
+/* COIN_UINT64_T */
+#cmakedefine COIN_UINT64_T ${COIN_UINT64_T}
+/* COIN_INTPTR_T */
+#cmakedefine COIN_INTPTR_T ${COIN_INTPTR_T}
+
+/* COIN_${NAME}_CHECKLEVEL */
+#define COIN_${NAME}_CHECKLEVEL ${COIN_${NAME}_CHECKLEVEL}
+/* COIN_${NAME}_VERBOSITY */
+#define COIN_${NAME}_VERBOSITY ${COIN_${NAME}_VERBOSITY}
+
+/* PACKAGE */
+#cmakedefine PACKAGE
+/* PACKAGE_NAME */
+#cmakedefine PACKAGE_NAME
+/* PACKAGE_VERSION */
+#cmakedefine PACKAGE_VERSION
+/* PACKAGE_STRING */
+#cmakedefine PACKAGE_STRING
+/* PACKAGE_TARNAME */
+#cmakedefine PACKAGE_TARNAME
+/* PACKAGE_BUGREPORT */
+#cmakedefine PACKAGE_BUGREPORT
+
diff --git a/Cbc/config_cbc.h.cmake.in b/Cbc/config_cbc.h.cmake.in
new file mode 100644
index 0000000..8dc44bf
--- /dev/null
+++ b/Cbc/config_cbc.h.cmake.in
@@ -0,0 +1,13 @@
+#ifndef __CONFIG_${NAME}_H__
+#define __CONFIG_${NAME}_H__
+
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+#endif
diff --git a/Cgl/CMakeLists.txt b/Cgl/CMakeLists.txt
new file mode 100644
index 0000000..2bd82cd
--- /dev/null
+++ b/Cgl/CMakeLists.txt
@@ -0,0 +1,151 @@
+message(STATUS "Subproject: Cgl")
+
+find_package(CoinUtils REQUIRED)
+find_package(Osi REQUIRED)
+find_package(OsiClp REQUIRED)
+
+set(NAME "CGL")
+parse_ac(VERSION MAJOR MINOR PATCH)
+
+set(COIN_${NAME}_CHECKLEVEL "0" CACHE INTERNAL
+	"${NAME} check level")
+set(COIN_${NAME}_VERBOSITY "0" CACHE INTERNAL
+	"${NAME} verbosity level")
+configure_file(config.h.cmake.in config.h)
+configure_file(config_cgl.h.cmake.in config_cgl.h)
+
+set(_SRCS
+	src/CglMixedIntegerRounding/CglMixedIntegerRoundingTest.cpp
+	src/CglMixedIntegerRounding/CglMixedIntegerRounding.cpp
+	src/CglDuplicateRow/CglDuplicateRow.cpp
+	src/CglCutGenerator.cpp
+	src/CglPreProcess/CglPreProcess.cpp
+	src/CglProbing/CglProbing.cpp
+	src/CglProbing/CglProbingTest.cpp
+	src/CglMixedIntegerRounding2/CglMixedIntegerRounding2.cpp
+	src/CglMixedIntegerRounding2/CglMixedIntegerRounding2Test.cpp
+	src/CglLandP/CglLandPUtils.cpp
+	src/CglLandP/CglLandPTest.cpp
+	src/CglLandP/CglLandPTabRow.cpp
+	src/CglLandP/CglLandPSimplex.cpp
+	src/CglLandP/CglLandPValidator.cpp
+	src/CglLandP/CglLandPMessages.cpp
+	src/CglLandP/CglLandP.cpp
+	src/CglOddHole/CglOddHoleTest.cpp
+	src/CglOddHole/CglOddHole.cpp
+	src/CglRedSplit/CglRedSplitParam.cpp
+	src/CglRedSplit/CglRedSplit.cpp
+	src/CglRedSplit/CglRedSplitTest.cpp
+	src/CglAllDifferent/CglAllDifferent.cpp
+	src/CglKnapsackCover/CglKnapsackCoverTest.cpp
+	src/CglKnapsackCover/CglKnapsackCover.cpp
+	src/CglSimpleRounding/CglSimpleRoundingTest.cpp
+	src/CglSimpleRounding/CglSimpleRounding.cpp
+	src/CglZeroHalf/CglZeroHalfTest.cpp
+	src/CglZeroHalf/Cgl012cut.cpp
+	src/CglZeroHalf/CglZeroHalf.cpp
+	src/CglRedSplit2/CglRedSplit2Test.cpp
+	src/CglRedSplit2/CglRedSplit2Param.cpp
+	src/CglRedSplit2/CglRedSplit2.cpp
+	src/CglMessage.cpp
+	src/CglStored.cpp
+	src/CglParam.cpp
+	src/CglResidualCapacity/CglResidualCapacity.cpp
+	src/CglResidualCapacity/CglResidualCapacityTest.cpp
+	src/CglTwomir/CglTwomirTest.cpp
+	src/CglTwomir/CglTwomir.cpp
+	src/CglFlowCover/CglFlowCoverTest.cpp
+	src/CglFlowCover/CglFlowCover.cpp
+	src/CglClique/CglCliqueHelper.cpp
+	src/CglClique/CglCliqueTest.cpp
+	src/CglClique/CglClique.cpp
+	src/CglTreeInfo.cpp
+	src/CglLiftAndProject/CglLiftAndProject.cpp
+	src/CglGomory/CglGomoryTest.cpp
+	src/CglGomory/CglGomory.cpp
+	src/CglGMI/CglGMI.cpp
+	src/CglGMI/CglGMIParam.cpp)
+
+set(_HDRS
+	src/CglConfig.h
+	src/CglMixedIntegerRounding/CglMixedIntegerRounding.hpp
+	src/CglDuplicateRow/CglDuplicateRow.hpp
+	src/CglStored.hpp
+	src/CglPreProcess/CglPreProcess.hpp
+	src/CglProbing/CglProbing.hpp
+	src/CglMixedIntegerRounding2/CglMixedIntegerRounding2.hpp
+	src/CglLandP/CglLandP.hpp
+	src/CglLandP/CglLandPUtils.hpp
+	src/CglLandP/CglLandPValidator.hpp
+	src/CglLandP/CglLandPTabRow.hpp
+	src/CglLandP/CglLandPMessages.hpp
+	src/CglLandP/CglLandPSimplex.hpp
+	src/CglOddHole/CglOddHole.hpp
+	src/CglRedSplit/CglRedSplitParam.hpp
+	src/CglRedSplit/CglRedSplit.hpp
+	src/CglAllDifferent/CglAllDifferent.hpp
+	src/CglKnapsackCover/CglKnapsackCover.hpp
+	src/CglSimpleRounding/CglSimpleRounding.hpp
+	src/CglMessage.hpp
+	src/CglZeroHalf/CglZeroHalf.hpp
+	src/CglZeroHalf/Cgl012cut.hpp
+	src/CglRedSplit2/CglRedSplit2Param.hpp
+	src/CglRedSplit2/CglRedSplit2.hpp
+	src/CglCutGenerator.hpp
+	src/CglResidualCapacity/CglResidualCapacity.hpp
+	src/CglTwomir/CglTwomir.hpp
+	src/CglFlowCover/CglFlowCover.hpp
+	src/CglClique/CglClique.hpp
+	src/CglLiftAndProject/CglLiftAndProject.hpp
+	src/CglGomory/CglGomory.hpp
+	src/CglGMI/CglGMIParam.hpp
+	src/CglGMI/CglGMI.hpp
+	src/CglParam.hpp
+	src/CglTreeInfo.hpp)
+
+add_library(Cgl ${_SRCS} ${_HDRS})
+target_include_directories(Cgl PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglClique>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglDuplicateRow>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglFlowCover>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglGMI>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglGomory>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglKnapsackCover>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglLandP>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglMixedIntegerRounding>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglMixedIntegerRounding2>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglPreProcess>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglProbing>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglOddHole>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglRedSplit>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglRedSplit2>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglResidualCapacity>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglSimpleRounding>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglTwomir>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/CglZeroHalf>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
+target_compile_definitions(Cgl
+	PUBLIC HAVE_CONFIG_H
+	PRIVATE CGL_BUILD)
+if(WIN32)
+	target_compile_definitions(Cgl PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(Cgl PUBLIC cxx_std_11)
+target_link_libraries(Cgl PUBLIC
+	CoinUtils::CoinUtils
+	Osi::Osi
+	OsiClp::OsiClp)
+set_target_properties(Cgl PROPERTIES
+	PUBLIC_HEADER "${_HDRS};${CMAKE_CURRENT_BINARY_DIR}/config_cgl.h"
+	VERSION ${VERSION})
+add_library(Cgl::Cgl ALIAS Cgl)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS Cgl
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
diff --git a/Cgl/config.h.cmake.in b/Cgl/config.h.cmake.in
new file mode 100644
index 0000000..35329dc
--- /dev/null
+++ b/Cgl/config.h.cmake.in
@@ -0,0 +1,114 @@
+/*config.h. Generated by configure_file.*/
+
+#define COIN_HAS_COINUTILS 1
+#define COIN_HAS_OSI 1
+#define COIN_HAS_OSICLP 1
+
+/* VERSION */
+#define VERSION "${VERSION}"
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+/* HAVE_MATH_H */
+#cmakedefine HAVE_MATH_H ${HAVE_MATH_H}
+/* HAVE_CTYPE_H */
+#cmakedefine HAVE_CTYPE_H ${HAVE_CTYPE_H}
+/* HAVE_INTTYPES_H */
+#cmakedefine HAVE_INTTYPES_H ${HAVE_INTTYPES_H}
+/* HAVE_FLOAT_H */
+#cmakedefine HAVE_FLOAT_H ${HAVE_FLOAT_H}
+/* HAVE_IEEEFP_H */
+#cmakedefine HAVE_IEEEFP_H ${HAVE_IEEEFP_H}
+/* HAVE_STDARG_H */
+#cmakedefine HAVE_STDARG_H ${HAVE_STDARG_H}
+/* HAVE_STDDEF_H */
+#cmakedefine HAVE_STDDEF_H ${HAVE_STDDEF_H}
+/* HAVE_STDINT_H */
+#cmakedefine HAVE_STDINT_H ${HAVE_STDINT_H}
+/* HAVE_STDIO_H */
+#cmakedefine HAVE_STDIO_H ${HAVE_STDIO_H}
+/* HAVE_STDLIB_H */
+#cmakedefine HAVE_STDLIB_H ${HAVE_STDLIB_H}
+/* HAVE_ASSERT_H */
+#cmakedefine HAVE_ASSERT_H ${HAVE_ASSERT_H}
+/* HAVE_DLFCN_H */
+#cmakedefine HAVE_DLFCN_H ${HAVE_DLFCN_H}
+/* HAVE_ENDIAN_H */
+#cmakedefine HAVE_ENDIAN_H ${HAVE_ENDIAN_H}
+/* HAVE_MEMORY_H */
+#cmakedefine HAVE_MEMORY_H ${HAVE_MEMORY_H}
+/* HAVE_STRINGS_H */
+#cmakedefine HAVE_STRINGS_H ${HAVE_STRINGS_H}
+/* HAVE_STRING_H */
+#cmakedefine HAVE_STRING_H ${HAVE_STRING_H}
+/* HAVE_TIME_H */
+#cmakedefine HAVE_TIME_H ${HAVE_TIME_H}
+/* HAVE_UNISTD_H */
+#cmakedefine HAVE_UNISTD_H ${HAVE_UNISTD_H}
+/* HAVE_SYS_STAT_H */
+#cmakedefine HAVE_SYS_STAT_H ${HAVE_SYS_STAT_H}
+/* HAVE_SYS_TYPES_H */
+#cmakedefine HAVE_SYS_TYPES_H ${HAVE_SYS_TYPES_H}
+
+/* HAVE_CMATH */
+#cmakedefine HAVE_CMATH ${HAVE_CMATH}
+/* HAVE_CCTYPE */
+#cmakedefine HAVE_CCTYPE ${HAVE_CCTYPE}
+/* HAVE_CINTTYPES */
+#cmakedefine HAVE_CINTTYPES ${HAVE_CINTTYPES}
+/* HAVE_CFLOAT */
+#cmakedefine HAVE_CFLOAT ${HAVE_CFLOAT}
+/* HAVE_CIEEEFP */
+#cmakedefine HAVE_CIEEEFP ${HAVE_CIEEEFP}
+/* HAVE_CSTDARG */
+#cmakedefine HAVE_CSTDARG ${HAVE_CSTDARG}
+/* HAVE_CSTDDEF */
+#cmakedefine HAVE_CSTDDEF ${HAVE_CSTDDEF}
+/* HAVE_CSTDINT */
+#cmakedefine HAVE_CSTDINT ${HAVE_CSTDINT}
+/* HAVE_CSTDIO */
+#cmakedefine HAVE_CSTDIO ${HAVE_CSTDIO}
+/* HAVE_CSTDLIB */
+#cmakedefine HAVE_CSTDLIB ${HAVE_CSTDLIB}
+/* HAVE_CASSERT */
+#cmakedefine HAVE_CASSERT ${HAVE_CASSERT}
+/* HAVE_CSTRING */
+#cmakedefine HAVE_CSTRING ${HAVE_CSTRING}
+/* HAVE_CTIME */
+#cmakedefine HAVE_CTIME ${HAVE_CTIME}
+
+/* COIN_C_FINITE */
+#cmakedefine COIN_C_FINITE ${COIN_C_FINITE}
+/* COIN_C_ISNAN */
+#cmakedefine COIN_C_ISNAN ${COIN_C_ISNAN}
+/* COIN_INT64_T */
+#cmakedefine COIN_INT64_T ${COIN_INT64_T}
+/* COIN_UINT64_T */
+#cmakedefine COIN_UINT64_T ${COIN_UINT64_T}
+/* COIN_INTPTR_T */
+#cmakedefine COIN_INTPTR_T ${COIN_INTPTR_T}
+
+/* COIN_${NAME}_CHECKLEVEL */
+#define COIN_${NAME}_CHECKLEVEL ${COIN_${NAME}_CHECKLEVEL}
+/* COIN_${NAME}_VERBOSITY */
+#define COIN_${NAME}_VERBOSITY ${COIN_${NAME}_VERBOSITY}
+
+/* PACKAGE */
+#cmakedefine PACKAGE
+/* PACKAGE_NAME */
+#cmakedefine PACKAGE_NAME
+/* PACKAGE_VERSION */
+#cmakedefine PACKAGE_VERSION
+/* PACKAGE_STRING */
+#cmakedefine PACKAGE_STRING
+/* PACKAGE_TARNAME */
+#cmakedefine PACKAGE_TARNAME
+/* PACKAGE_BUGREPORT */
+#cmakedefine PACKAGE_BUGREPORT
+
diff --git a/Cgl/config_cgl.h.cmake.in b/Cgl/config_cgl.h.cmake.in
new file mode 100644
index 0000000..8dc44bf
--- /dev/null
+++ b/Cgl/config_cgl.h.cmake.in
@@ -0,0 +1,13 @@
+#ifndef __CONFIG_${NAME}_H__
+#define __CONFIG_${NAME}_H__
+
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+#endif
diff --git a/Clp/CMakeLists.txt b/Clp/CMakeLists.txt
new file mode 100644
index 0000000..6f7ff7c
--- /dev/null
+++ b/Clp/CMakeLists.txt
@@ -0,0 +1,208 @@
+message(STATUS "Subproject: Clp")
+
+find_package(CoinUtils REQUIRED)
+find_package(Osi REQUIRED)
+
+set(NAME "CLP")
+parse_ac(VERSION MAJOR MINOR PATCH)
+
+set(COIN_${NAME}_CHECKLEVEL "0" CACHE INTERNAL
+	"${NAME} check level")
+set(COIN_${NAME}_VERBOSITY "0" CACHE INTERNAL
+	"${NAME} verbosity level")
+configure_file(config.h.cmake.in config.h)
+configure_file(config_clp.h.cmake.in config_clp.h)
+
+set(_SRCS
+	src/ClpCholeskyBase.cpp
+	src/ClpCholeskyDense.cpp
+	src/ClpConstraint.cpp
+	src/ClpConstraintLinear.cpp
+	src/ClpConstraintQuadratic.cpp
+	src/Clp_C_Interface.cpp
+	src/ClpDualRowDantzig.cpp
+	src/ClpDualRowPivot.cpp
+	src/ClpDualRowSteepest.cpp
+	src/ClpDummyMatrix.cpp
+	src/ClpDynamicExampleMatrix.cpp
+	src/ClpDynamicMatrix.cpp
+	src/ClpEventHandler.cpp
+	src/ClpFactorization.cpp
+	src/ClpGubDynamicMatrix.cpp
+	src/ClpGubMatrix.cpp
+	src/ClpHelperFunctions.cpp
+	src/ClpInterior.cpp
+	src/ClpLinearObjective.cpp
+	src/ClpMatrixBase.cpp
+	src/ClpMessage.cpp
+	src/ClpModel.cpp
+	src/ClpNetworkBasis.cpp
+	src/ClpNetworkMatrix.cpp
+	src/ClpNonLinearCost.cpp
+	src/ClpNode.cpp
+	src/ClpObjective.cpp
+	src/ClpPackedMatrix.cpp
+	src/ClpPlusMinusOneMatrix.cpp
+	src/ClpPredictorCorrector.cpp
+	src/ClpPdco.cpp
+	src/ClpPdcoBase.cpp
+	src/ClpLsqr.cpp
+	src/ClpPresolve.cpp
+	src/ClpPrimalColumnDantzig.cpp
+	src/ClpPrimalColumnPivot.cpp
+	src/ClpPrimalColumnSteepest.cpp
+	src/ClpQuadraticObjective.cpp
+	src/ClpSimplex.cpp
+	src/ClpSimplexDual.cpp
+	src/ClpSimplexNonlinear.cpp
+	src/ClpSimplexOther.cpp
+	src/ClpSimplexPrimal.cpp
+	src/ClpSolve.cpp
+	src/Idiot.cpp
+	src/IdiSolve.cpp)
+
+set(_HDRS
+	src/Clp_C_Interface.h
+	src/ClpConfig.h
+	src/ClpCholeskyBase.hpp
+	src/ClpCholeskyDense.hpp
+	src/ClpConstraint.hpp
+	src/ClpConstraintLinear.hpp
+	src/ClpConstraintQuadratic.hpp
+	src/ClpDualRowDantzig.hpp
+	src/ClpDualRowPivot.hpp
+	src/ClpDualRowSteepest.hpp
+	src/ClpDummyMatrix.hpp
+	src/ClpDynamicExampleMatrix.hpp
+	src/ClpDynamicMatrix.hpp
+	src/ClpEventHandler.hpp
+	src/ClpFactorization.hpp
+	src/ClpGubDynamicMatrix.hpp
+	src/ClpGubMatrix.hpp
+	src/ClpInterior.hpp
+	src/ClpLinearObjective.hpp
+	src/ClpMatrixBase.hpp
+	src/ClpMessage.hpp
+	src/ClpModel.hpp
+	src/ClpNetworkMatrix.hpp
+	src/ClpNonLinearCost.hpp
+	src/ClpNode.hpp
+	src/ClpObjective.hpp
+	src/ClpPackedMatrix.hpp
+	src/ClpPdcoBase.hpp
+	src/ClpPlusMinusOneMatrix.hpp
+	src/ClpParameters.hpp
+	src/ClpPresolve.hpp
+	src/ClpPrimalColumnDantzig.hpp
+	src/ClpPrimalColumnPivot.hpp
+	src/ClpPrimalColumnSteepest.hpp
+	src/ClpQuadraticObjective.hpp
+	src/ClpSimplex.hpp
+	src/ClpSimplexNonlinear.hpp
+	src/ClpSimplexOther.hpp
+	src/ClpSimplexDual.hpp
+	src/ClpSimplexPrimal.hpp
+	src/ClpSolve.hpp
+	src/CbcOrClpParam.hpp
+	src/Idiot.hpp)
+
+add_library(Clp ${_SRCS} ${_HDRS})
+target_include_directories(Clp PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+	$<INSTALL_INTERFACE:include/coin>)
+target_compile_definitions(Clp
+	PUBLIC HAVE_CONFIG_H
+	PRIVATE CLP_BUILD)
+if(WIN32)
+	target_compile_definitions(Clp PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(Clp PUBLIC cxx_std_11)
+target_link_libraries(Clp PUBLIC
+	CoinUtils::CoinUtils
+	Osi::Osi)
+set_target_properties(Clp PROPERTIES
+	PUBLIC_HEADER "${_HDRS};${CMAKE_CURRENT_BINARY_DIR}/config_clp.h"
+	VERSION ${VERSION})
+add_library(Clp::Clp ALIAS Clp)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS Clp
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
+
+##############
+##  OsiClp  ##
+##############
+set(_SRCS src/OsiClp/OsiClpSolverInterface.cpp)
+set(_HDRS	src/OsiClp/OsiClpSolverInterface.hpp)
+
+add_library(OsiClp ${_SRCS} ${_HDRS})
+target_include_directories(OsiClp PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/OsiClp>)
+#target_compile_definitions(OsiClp PRIVATE ...)
+if(WIN32)
+	target_compile_definitions(OsiClp PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(OsiClp PUBLIC cxx_std_11)
+target_link_libraries(OsiClp PUBLIC
+	CoinUtils::CoinUtils
+	Osi::Osi
+	Clp::Clp)
+set_target_properties(OsiClp PROPERTIES
+	PUBLIC_HEADER "${_HDRS}"
+	VERSION ${VERSION})
+add_library(OsiClp::OsiClp ALIAS OsiClp)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS OsiClp
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
+
+#################
+##  ClpSolver  ##
+#################
+set(_SRCS
+	src/ClpSolver.cpp
+	src/CbcOrClpParam.cpp
+	src/Clp_ampl.cpp
+	src/MyEventHandler.cpp
+	src/MyMessageHandler.cpp)
+
+set(_HDRS
+	src/CbcOrClpParam.hpp
+	src/Clp_ampl.h
+	src/MyEventHandler.hpp
+	src/MyMessageHandler.hpp)
+
+add_library(ClpSolver ${_SRCS} ${_HDRS})
+target_include_directories(ClpSolver PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>)
+#target_compile_definitions(ClpSolver PRIVATE ...)
+if(WIN32)
+	target_compile_definitions(ClpSolver PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(ClpSolver PUBLIC cxx_std_11)
+target_link_libraries(ClpSolver PUBLIC
+	Clp::Clp)
+set_target_properties(ClpSolver PROPERTIES
+	PUBLIC_HEADER "${_HDRS}"
+	VERSION ${VERSION})
+add_library(ClpSolver::ClpSolver ALIAS ClpSolver)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS ClpSolver
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
diff --git a/Clp/config.h.cmake.in b/Clp/config.h.cmake.in
new file mode 100644
index 0000000..c12b3d4
--- /dev/null
+++ b/Clp/config.h.cmake.in
@@ -0,0 +1,113 @@
+/*config.h. Generated by configure_file.*/
+
+#define COIN_HAS_COINUTILS 1
+#define COIN_HAS_OSI 1
+
+/* VERSION */
+#define VERSION "${VERSION}"
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+/* HAVE_MATH_H */
+#cmakedefine HAVE_MATH_H ${HAVE_MATH_H}
+/* HAVE_CTYPE_H */
+#cmakedefine HAVE_CTYPE_H ${HAVE_CTYPE_H}
+/* HAVE_INTTYPES_H */
+#cmakedefine HAVE_INTTYPES_H ${HAVE_INTTYPES_H}
+/* HAVE_FLOAT_H */
+#cmakedefine HAVE_FLOAT_H ${HAVE_FLOAT_H}
+/* HAVE_IEEEFP_H */
+#cmakedefine HAVE_IEEEFP_H ${HAVE_IEEEFP_H}
+/* HAVE_STDARG_H */
+#cmakedefine HAVE_STDARG_H ${HAVE_STDARG_H}
+/* HAVE_STDDEF_H */
+#cmakedefine HAVE_STDDEF_H ${HAVE_STDDEF_H}
+/* HAVE_STDINT_H */
+#cmakedefine HAVE_STDINT_H ${HAVE_STDINT_H}
+/* HAVE_STDIO_H */
+#cmakedefine HAVE_STDIO_H ${HAVE_STDIO_H}
+/* HAVE_STDLIB_H */
+#cmakedefine HAVE_STDLIB_H ${HAVE_STDLIB_H}
+/* HAVE_ASSERT_H */
+#cmakedefine HAVE_ASSERT_H ${HAVE_ASSERT_H}
+/* HAVE_DLFCN_H */
+#cmakedefine HAVE_DLFCN_H ${HAVE_DLFCN_H}
+/* HAVE_ENDIAN_H */
+#cmakedefine HAVE_ENDIAN_H ${HAVE_ENDIAN_H}
+/* HAVE_MEMORY_H */
+#cmakedefine HAVE_MEMORY_H ${HAVE_MEMORY_H}
+/* HAVE_STRINGS_H */
+#cmakedefine HAVE_STRINGS_H ${HAVE_STRINGS_H}
+/* HAVE_STRING_H */
+#cmakedefine HAVE_STRING_H ${HAVE_STRING_H}
+/* HAVE_TIME_H */
+#cmakedefine HAVE_TIME_H ${HAVE_TIME_H}
+/* HAVE_UNISTD_H */
+#cmakedefine HAVE_UNISTD_H ${HAVE_UNISTD_H}
+/* HAVE_SYS_STAT_H */
+#cmakedefine HAVE_SYS_STAT_H ${HAVE_SYS_STAT_H}
+/* HAVE_SYS_TYPES_H */
+#cmakedefine HAVE_SYS_TYPES_H ${HAVE_SYS_TYPES_H}
+
+/* HAVE_CMATH */
+#cmakedefine HAVE_CMATH ${HAVE_CMATH}
+/* HAVE_CCTYPE */
+#cmakedefine HAVE_CCTYPE ${HAVE_CCTYPE}
+/* HAVE_CINTTYPES */
+#cmakedefine HAVE_CINTTYPES ${HAVE_CINTTYPES}
+/* HAVE_CFLOAT */
+#cmakedefine HAVE_CFLOAT ${HAVE_CFLOAT}
+/* HAVE_CIEEEFP */
+#cmakedefine HAVE_CIEEEFP ${HAVE_CIEEEFP}
+/* HAVE_CSTDARG */
+#cmakedefine HAVE_CSTDARG ${HAVE_CSTDARG}
+/* HAVE_CSTDDEF */
+#cmakedefine HAVE_CSTDDEF ${HAVE_CSTDDEF}
+/* HAVE_CSTDINT */
+#cmakedefine HAVE_CSTDINT ${HAVE_CSTDINT}
+/* HAVE_CSTDIO */
+#cmakedefine HAVE_CSTDIO ${HAVE_CSTDIO}
+/* HAVE_CSTDLIB */
+#cmakedefine HAVE_CSTDLIB ${HAVE_CSTDLIB}
+/* HAVE_CASSERT */
+#cmakedefine HAVE_CASSERT ${HAVE_CASSERT}
+/* HAVE_CSTRING */
+#cmakedefine HAVE_CSTRING ${HAVE_CSTRING}
+/* HAVE_CTIME */
+#cmakedefine HAVE_CTIME ${HAVE_CTIME}
+
+/* COIN_C_FINITE */
+#cmakedefine COIN_C_FINITE ${COIN_C_FINITE}
+/* COIN_C_ISNAN */
+#cmakedefine COIN_C_ISNAN ${COIN_C_ISNAN}
+/* COIN_INT64_T */
+#cmakedefine COIN_INT64_T ${COIN_INT64_T}
+/* COIN_UINT64_T */
+#cmakedefine COIN_UINT64_T ${COIN_UINT64_T}
+/* COIN_INTPTR_T */
+#cmakedefine COIN_INTPTR_T ${COIN_INTPTR_T}
+
+/* COIN_${NAME}_CHECKLEVEL */
+#define COIN_${NAME}_CHECKLEVEL ${COIN_${NAME}_CHECKLEVEL}
+/* COIN_${NAME}_VERBOSITY */
+#define COIN_${NAME}_VERBOSITY ${COIN_${NAME}_VERBOSITY}
+
+/* PACKAGE */
+#cmakedefine PACKAGE
+/* PACKAGE_NAME */
+#cmakedefine PACKAGE_NAME
+/* PACKAGE_VERSION */
+#cmakedefine PACKAGE_VERSION
+/* PACKAGE_STRING */
+#cmakedefine PACKAGE_STRING
+/* PACKAGE_TARNAME */
+#cmakedefine PACKAGE_TARNAME
+/* PACKAGE_BUGREPORT */
+#cmakedefine PACKAGE_BUGREPORT
+
diff --git a/Clp/config_clp.h.cmake.in b/Clp/config_clp.h.cmake.in
new file mode 100644
index 0000000..8dc44bf
--- /dev/null
+++ b/Clp/config_clp.h.cmake.in
@@ -0,0 +1,13 @@
+#ifndef __CONFIG_${NAME}_H__
+#define __CONFIG_${NAME}_H__
+
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+#endif
diff --git a/CoinUtils/CMakeLists.txt b/CoinUtils/CMakeLists.txt
new file mode 100644
index 0000000..c2888e7
--- /dev/null
+++ b/CoinUtils/CMakeLists.txt
@@ -0,0 +1,174 @@
+message(STATUS "Subproject: CoinUtils")
+
+set(NAME "COINUTILS")
+parse_ac(VERSION MAJOR MINOR PATCH)
+
+# PTHREAD
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+find_package(Threads)
+if(CMAKE_USE_PTHREADS_INIT)
+	set(${NAME}_PTHREADS "1" CACHE INTERNAL "Use pthread")
+endif()
+
+set(COIN_${NAME}_CHECKLEVEL "0" CACHE INTERNAL
+	"${NAME} check level")
+set(COIN_${NAME}_VERBOSITY "0" CACHE INTERNAL
+	"${NAME} verbosity level")
+configure_file(config.h.cmake.in config.h)
+configure_file(config_coinutils.h.cmake.in config_coinutils.h)
+
+set(_SRCS
+	src/CoinAlloc.cpp
+	src/CoinBuild.cpp
+	src/CoinDenseVector.cpp
+	src/CoinError.cpp
+	src/CoinFactorization1.cpp
+	src/CoinFactorization2.cpp
+	src/CoinFactorization3.cpp
+	src/CoinFactorization4.cpp
+	src/CoinSimpFactorization.cpp
+	src/CoinDenseFactorization.cpp
+	src/CoinOslFactorization.cpp
+	src/CoinOslFactorization2.cpp
+	src/CoinOslFactorization3.cpp
+	src/CoinFileIO.cpp
+	src/CoinFinite.cpp
+	src/CoinIndexedVector.cpp
+	src/CoinLpIO.cpp
+	src/CoinMessage.cpp
+	src/CoinMessageHandler.cpp
+	src/CoinModel.cpp
+	src/CoinStructuredModel.cpp
+	src/CoinModelUseful.cpp
+	src/CoinModelUseful2.cpp
+	src/CoinMpsIO.cpp
+	src/CoinPackedMatrix.cpp
+	src/CoinPackedVector.cpp
+	src/CoinPackedVectorBase.cpp
+	src/CoinParam.cpp
+	src/CoinParamUtils.cpp
+	src/CoinPostsolveMatrix.cpp
+	src/CoinPrePostsolveMatrix.cpp
+	src/CoinPresolveDoubleton.cpp
+	src/CoinPresolveDual.cpp
+	src/CoinPresolveDupcol.cpp
+	src/CoinPresolveEmpty.cpp
+	src/CoinPresolveFixed.cpp
+	src/CoinPresolveForcing.cpp
+	src/CoinPresolveHelperFunctions.cpp
+	src/CoinPresolveImpliedFree.cpp
+	src/CoinPresolveIsolated.cpp
+	src/CoinPresolveMatrix.cpp
+	src/CoinPresolvePsdebug.cpp
+	src/CoinPresolveMonitor.cpp
+	src/CoinPresolveSingleton.cpp
+	src/CoinPresolveSubst.cpp
+	src/CoinPresolveTighten.cpp
+	src/CoinPresolveTripleton.cpp
+	src/CoinPresolveUseless.cpp
+	src/CoinPresolveZeros.cpp
+	src/CoinSearchTree.cpp
+	src/CoinShallowPackedVector.cpp
+	src/CoinSnapshot.cpp
+	src/CoinWarmStartBasis.cpp
+	src/CoinWarmStartVector.cpp
+	src/CoinWarmStartDual.cpp
+	src/CoinWarmStartPrimalDual.cpp
+	src/CoinRational.cpp)
+
+set(_HDRS
+	src/Coin_C_defines.h
+	src/CoinUtilsConfig.h
+	src/CoinAlloc.hpp
+	src/CoinBuild.hpp
+	src/CoinDenseVector.hpp
+	src/CoinDistance.hpp
+	src/CoinError.hpp
+	src/CoinFactorization.hpp
+	src/CoinSimpFactorization.hpp
+	src/CoinDenseFactorization.hpp
+	src/CoinOslFactorization.hpp
+	src/CoinFileIO.hpp
+	src/CoinFinite.hpp
+	src/CoinFloatEqual.hpp
+	src/CoinHelperFunctions.hpp
+	src/CoinIndexedVector.hpp
+	src/CoinLpIO.hpp
+	src/CoinMessage.hpp
+	src/CoinMessageHandler.hpp
+	src/CoinModel.hpp
+	src/CoinStructuredModel.hpp
+	src/CoinModelUseful.hpp
+	src/CoinMpsIO.hpp
+	src/CoinPackedMatrix.hpp
+	src/CoinPackedVector.hpp
+	src/CoinPackedVectorBase.hpp
+	src/CoinParam.hpp
+	src/CoinPragma.hpp
+	src/CoinPresolveDoubleton.hpp
+	src/CoinPresolveDual.hpp
+	src/CoinPresolveDupcol.hpp
+	src/CoinPresolveEmpty.hpp
+	src/CoinPresolveFixed.hpp
+	src/CoinPresolveForcing.hpp
+	src/CoinPresolveImpliedFree.hpp
+	src/CoinPresolveIsolated.hpp
+	src/CoinPresolveMatrix.hpp
+	src/CoinPresolveMonitor.hpp
+	src/CoinPresolvePsdebug.hpp
+	src/CoinPresolveSingleton.hpp
+	src/CoinPresolveSubst.hpp
+	src/CoinPresolveTighten.hpp
+	src/CoinPresolveTripleton.hpp
+	src/CoinPresolveUseless.hpp
+	src/CoinPresolveZeros.hpp
+	src/CoinSearchTree.hpp
+	src/CoinShallowPackedVector.hpp
+	src/CoinSignal.hpp
+	src/CoinSmartPtr.hpp
+	src/CoinSnapshot.hpp
+	src/CoinSort.hpp
+	src/CoinTime.hpp
+	src/CoinTypes.hpp
+	src/CoinUtility.hpp
+	src/CoinWarmStart.hpp
+	src/CoinWarmStartBasis.hpp
+	src/CoinWarmStartVector.hpp
+	src/CoinWarmStartDual.hpp
+	src/CoinWarmStartPrimalDual.hpp
+	src/CoinRational.hpp)
+
+add_library(CoinUtils ${_SRCS} ${_HDRS})
+target_include_directories(CoinUtils PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+	$<INSTALL_INTERFACE:include/coin>)
+target_compile_definitions(CoinUtils
+	PUBLIC HAVE_CONFIG_H
+	PRIVATE COINUTILS_BUILD)
+if(WIN32)
+	target_compile_definitions(CoinUtils PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(CoinUtils PUBLIC cxx_std_11)
+
+if(ZLIB_FOUND)
+	target_link_libraries(CoinUtils PRIVATE	ZLIB::ZLIB)
+endif()
+
+if(CMAKE_USE_PTHREADS_INIT)
+	target_link_libraries(CoinUtils PRIVATE	Threads::Threads)
+endif()
+
+set_target_properties(CoinUtils PROPERTIES
+	PUBLIC_HEADER "${_HDRS};${CMAKE_CURRENT_BINARY_DIR}/config_coinutils.h"
+	VERSION ${VERSION})
+add_library(CoinUtils::CoinUtils ALIAS CoinUtils)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS CoinUtils
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
diff --git a/CoinUtils/config.h.cmake.in b/CoinUtils/config.h.cmake.in
new file mode 100644
index 0000000..ecc0a94
--- /dev/null
+++ b/CoinUtils/config.h.cmake.in
@@ -0,0 +1,127 @@
+/*config.h. Generated by configure_file.*/
+
+/* VERSION */
+#define VERSION "${VERSION}"
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+/* HAVE_MATH_H */
+#cmakedefine HAVE_MATH_H ${HAVE_MATH_H}
+/* HAVE_CTYPE_H */
+#cmakedefine HAVE_CTYPE_H ${HAVE_CTYPE_H}
+/* HAVE_INTTYPES_H */
+#cmakedefine HAVE_INTTYPES_H ${HAVE_INTTYPES_H}
+/* HAVE_FLOAT_H */
+#cmakedefine HAVE_FLOAT_H ${HAVE_FLOAT_H}
+/* HAVE_IEEEFP_H */
+#cmakedefine HAVE_IEEEFP_H ${HAVE_IEEEFP_H}
+/* HAVE_STDARG_H */
+#cmakedefine HAVE_STDARG_H ${HAVE_STDARG_H}
+/* HAVE_STDDEF_H */
+#cmakedefine HAVE_STDDEF_H ${HAVE_STDDEF_H}
+/* HAVE_STDINT_H */
+#cmakedefine HAVE_STDINT_H ${HAVE_STDINT_H}
+/* HAVE_STDIO_H */
+#cmakedefine HAVE_STDIO_H ${HAVE_STDIO_H}
+/* HAVE_STDLIB_H */
+#cmakedefine HAVE_STDLIB_H ${HAVE_STDLIB_H}
+/* HAVE_ASSERT_H */
+#cmakedefine HAVE_ASSERT_H ${HAVE_ASSERT_H}
+/* HAVE_DLFCN_H */
+#cmakedefine HAVE_DLFCN_H ${HAVE_DLFCN_H}
+/* HAVE_ENDIAN_H */
+#cmakedefine HAVE_ENDIAN_H ${HAVE_ENDIAN_H}
+/* HAVE_MEMORY_H */
+#cmakedefine HAVE_MEMORY_H ${HAVE_MEMORY_H}
+/* HAVE_STRINGS_H */
+#cmakedefine HAVE_STRINGS_H ${HAVE_STRINGS_H}
+/* HAVE_STRING_H */
+#cmakedefine HAVE_STRING_H ${HAVE_STRING_H}
+/* HAVE_TIME_H */
+#cmakedefine HAVE_TIME_H ${HAVE_TIME_H}
+/* HAVE_UNISTD_H */
+#cmakedefine HAVE_UNISTD_H ${HAVE_UNISTD_H}
+/* HAVE_SYS_STAT_H */
+#cmakedefine HAVE_SYS_STAT_H ${HAVE_SYS_STAT_H}
+/* HAVE_SYS_TYPES_H */
+#cmakedefine HAVE_SYS_TYPES_H ${HAVE_SYS_TYPES_H}
+
+/* HAVE_CMATH */
+#cmakedefine HAVE_CMATH ${HAVE_CMATH}
+/* HAVE_CCTYPE */
+#cmakedefine HAVE_CCTYPE ${HAVE_CCTYPE}
+/* HAVE_CINTTYPES */
+#cmakedefine HAVE_CINTTYPES ${HAVE_CINTTYPES}
+/* HAVE_CFLOAT */
+#cmakedefine HAVE_CFLOAT ${HAVE_CFLOAT}
+/* HAVE_CIEEEFP */
+#cmakedefine HAVE_CIEEEFP ${HAVE_CIEEEFP}
+/* HAVE_CSTDARG */
+#cmakedefine HAVE_CSTDARG ${HAVE_CSTDARG}
+/* HAVE_CSTDDEF */
+#cmakedefine HAVE_CSTDDEF ${HAVE_CSTDDEF}
+/* HAVE_CSTDINT */
+#cmakedefine HAVE_CSTDINT ${HAVE_CSTDINT}
+/* HAVE_CSTDIO */
+#cmakedefine HAVE_CSTDIO ${HAVE_CSTDIO}
+/* HAVE_CSTDLIB */
+#cmakedefine HAVE_CSTDLIB ${HAVE_CSTDLIB}
+/* HAVE_CASSERT */
+#cmakedefine HAVE_CASSERT ${HAVE_CASSERT}
+/* HAVE_CSTRING */
+#cmakedefine HAVE_CSTRING ${HAVE_CSTRING}
+/* HAVE_CTIME */
+#cmakedefine HAVE_CTIME ${HAVE_CTIME}
+
+/* STDC_HEADERS */
+#cmakedefine STDC_HEADERS ${STDC_HEADERS}
+/* COIN_C_FINITE */
+#cmakedefine COIN_C_FINITE ${COIN_C_FINITE}
+/* COIN_C_ISNAN */
+#cmakedefine COIN_C_ISNAN ${COIN_C_ISNAN}
+/* COIN_INT64_T */
+#cmakedefine COIN_INT64_T ${COIN_INT64_T}
+/* COIN_UINT64_T */
+#cmakedefine COIN_UINT64_T ${COIN_UINT64_T}
+/* COIN_INTPTR_T */
+#cmakedefine COIN_INTPTR_T ${COIN_INTPTR_T}
+
+/* COIN_${NAME}_CHECKLEVEL */
+#define COIN_${NAME}_CHECKLEVEL ${COIN_${NAME}_CHECKLEVEL}
+/* COIN_${NAME}_VERBOSITY */
+#define COIN_${NAME}_VERBOSITY ${COIN_${NAME}_VERBOSITY}
+
+/* ${NAME}_HAS_CSTDINT */
+#ifdef HAVE_CSTDINT
+  #define ${NAME}_HAS_CSTDINT 1
+#endif
+
+/* ${NAME}_PTHREADS */
+#cmakedefine COINUTILS_PTHREADS ${COINUTILS_PTHREADS}
+/* ${NAME}_MEMPOOL_MAXPOOLED */
+#define ${NAME}_MEMPOOL_MAXPOOLED -1
+
+/* HAVE_ZLIB_H */
+#cmakedefine HAVE_ZLIB_H ${HAVE_ZLIB_H}
+/* COIN_HAS_ZLIB */
+#cmakedefine COIN_HAS_ZLIB ${COIN_HAS_ZLIB}
+
+/* PACKAGE */
+#cmakedefine PACKAGE
+/* PACKAGE_NAME */
+#cmakedefine PACKAGE_NAME
+/* PACKAGE_VERSION */
+#cmakedefine PACKAGE_VERSION
+/* PACKAGE_STRING */
+#cmakedefine PACKAGE_STRING
+/* PACKAGE_TARNAME */
+#cmakedefine PACKAGE_TARNAME
+/* PACKAGE_BUGREPORT */
+#cmakedefine PACKAGE_BUGREPORT
+
diff --git a/CoinUtils/config_coinutils.h.cmake.in b/CoinUtils/config_coinutils.h.cmake.in
new file mode 100644
index 0000000..0d85fd7
--- /dev/null
+++ b/CoinUtils/config_coinutils.h.cmake.in
@@ -0,0 +1,22 @@
+#ifndef __CONFIG_${NAME}_H__
+#define __CONFIG_${NAME}_H__
+
+#define ${NAME}_HAS_CSTDINT 1
+
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+/* COIN_INT64_T */
+#cmakedefine COIN_INT64_T ${COIN_INT64_T}
+/* COIN_UINT64_T */
+#cmakedefine COIN_UINT64_T ${COIN_UINT64_T}
+/* COIN_INTPTR_T */
+#cmakedefine COIN_INTPTR_T ${COIN_INTPTR_T}
+
+#endif
diff --git a/Osi/CMakeLists.txt b/Osi/CMakeLists.txt
new file mode 100644
index 0000000..2d06bfd
--- /dev/null
+++ b/Osi/CMakeLists.txt
@@ -0,0 +1,71 @@
+message(STATUS "Subproject: Osi")
+
+find_package(CoinUtils REQUIRED)
+
+set(NAME "OSI")
+parse_ac(VERSION MAJOR MINOR PATCH)
+
+set(COIN_${NAME}_CHECKLEVEL "0" CACHE INTERNAL
+	"${NAME} check level")
+set(COIN_${NAME}_VERBOSITY "0" CACHE INTERNAL
+	"${NAME} verbosity level")
+configure_file(config.h.cmake.in config.h)
+configure_file(config_osi.h.cmake.in config_osi.h)
+
+set(_SRCS
+	src/Osi/OsiAuxInfo.cpp
+	src/Osi/OsiBranchingObject.cpp
+	src/Osi/OsiChooseVariable.cpp
+	src/Osi/OsiColCut.cpp
+	src/Osi/OsiCut.cpp
+	src/Osi/OsiCuts.cpp
+	src/Osi/OsiNames.cpp
+	src/Osi/OsiPresolve.cpp
+	src/Osi/OsiRowCut.cpp
+	src/Osi/OsiRowCutDebugger.cpp
+	src/Osi/OsiSolverBranch.cpp
+	src/Osi/OsiSolverInterface.cpp)
+
+set(_HDRS
+	src/Osi/OsiConfig.h
+	src/Osi/OsiAuxInfo.hpp
+	src/Osi/OsiBranchingObject.hpp
+	src/Osi/OsiChooseVariable.hpp
+	src/Osi/OsiColCut.hpp
+	src/Osi/OsiCollections.hpp
+	src/Osi/OsiCut.hpp
+	src/Osi/OsiCuts.hpp
+	src/Osi/OsiPresolve.hpp
+	src/Osi/OsiRowCut.hpp
+	src/Osi/OsiRowCutDebugger.hpp
+	src/Osi/OsiSolverBranch.hpp
+	src/Osi/OsiSolverInterface.hpp
+	src/Osi/OsiSolverParameters.hpp)
+
+add_library(Osi ${_SRCS} ${_HDRS})
+target_include_directories(Osi PUBLIC
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/Osi>
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+	$<INSTALL_INTERFACE:include/coin>)
+target_compile_definitions(Osi
+	PUBLIC HAVE_CONFIG_H
+	PRIVATE OSI_BUILD)
+if(WIN32)
+	target_compile_definitions(Osi PRIVATE _CRT_SECURE_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE)
+endif()
+target_compile_features(Osi PUBLIC cxx_std_11)
+target_link_libraries(Osi PUBLIC
+	CoinUtils::CoinUtils)
+set_target_properties(Osi PROPERTIES
+	PUBLIC_HEADER "${_HDRS};${CMAKE_CURRENT_BINARY_DIR}/config_osi.h"
+	VERSION ${VERSION})
+add_library(Osi::Osi ALIAS Osi)
+
+# Install
+include(GNUInstallDirs)
+install(TARGETS Osi
+	INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/coin/
+	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	)
diff --git a/Osi/config.h.cmake.in b/Osi/config.h.cmake.in
new file mode 100644
index 0000000..f544df8
--- /dev/null
+++ b/Osi/config.h.cmake.in
@@ -0,0 +1,112 @@
+/*config.h. Generated by configure_file.*/
+
+#define COIN_HAS_COINUTILS 1
+
+/* VERSION */
+#define VERSION "${VERSION}"
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+/* HAVE_MATH_H */
+#cmakedefine HAVE_MATH_H ${HAVE_MATH_H}
+/* HAVE_CTYPE_H */
+#cmakedefine HAVE_CTYPE_H ${HAVE_CTYPE_H}
+/* HAVE_INTTYPES_H */
+#cmakedefine HAVE_INTTYPES_H ${HAVE_INTTYPES_H}
+/* HAVE_FLOAT_H */
+#cmakedefine HAVE_FLOAT_H ${HAVE_FLOAT_H}
+/* HAVE_IEEEFP_H */
+#cmakedefine HAVE_IEEEFP_H ${HAVE_IEEEFP_H}
+/* HAVE_STDARG_H */
+#cmakedefine HAVE_STDARG_H ${HAVE_STDARG_H}
+/* HAVE_STDDEF_H */
+#cmakedefine HAVE_STDDEF_H ${HAVE_STDDEF_H}
+/* HAVE_STDINT_H */
+#cmakedefine HAVE_STDINT_H ${HAVE_STDINT_H}
+/* HAVE_STDIO_H */
+#cmakedefine HAVE_STDIO_H ${HAVE_STDIO_H}
+/* HAVE_STDLIB_H */
+#cmakedefine HAVE_STDLIB_H ${HAVE_STDLIB_H}
+/* HAVE_ASSERT_H */
+#cmakedefine HAVE_ASSERT_H ${HAVE_ASSERT_H}
+/* HAVE_DLFCN_H */
+#cmakedefine HAVE_DLFCN_H ${HAVE_DLFCN_H}
+/* HAVE_ENDIAN_H */
+#cmakedefine HAVE_ENDIAN_H ${HAVE_ENDIAN_H}
+/* HAVE_MEMORY_H */
+#cmakedefine HAVE_MEMORY_H ${HAVE_MEMORY_H}
+/* HAVE_STRINGS_H */
+#cmakedefine HAVE_STRINGS_H ${HAVE_STRINGS_H}
+/* HAVE_STRING_H */
+#cmakedefine HAVE_STRING_H ${HAVE_STRING_H}
+/* HAVE_TIME_H */
+#cmakedefine HAVE_TIME_H ${HAVE_TIME_H}
+/* HAVE_UNISTD_H */
+#cmakedefine HAVE_UNISTD_H ${HAVE_UNISTD_H}
+/* HAVE_SYS_STAT_H */
+#cmakedefine HAVE_SYS_STAT_H ${HAVE_SYS_STAT_H}
+/* HAVE_SYS_TYPES_H */
+#cmakedefine HAVE_SYS_TYPES_H ${HAVE_SYS_TYPES_H}
+
+/* HAVE_CMATH */
+#cmakedefine HAVE_CMATH ${HAVE_CMATH}
+/* HAVE_CCTYPE */
+#cmakedefine HAVE_CCTYPE ${HAVE_CCTYPE}
+/* HAVE_CINTTYPES */
+#cmakedefine HAVE_CINTTYPES ${HAVE_CINTTYPES}
+/* HAVE_CFLOAT */
+#cmakedefine HAVE_CFLOAT ${HAVE_CFLOAT}
+/* HAVE_CIEEEFP */
+#cmakedefine HAVE_CIEEEFP ${HAVE_CIEEEFP}
+/* HAVE_CSTDARG */
+#cmakedefine HAVE_CSTDARG ${HAVE_CSTDARG}
+/* HAVE_CSTDDEF */
+#cmakedefine HAVE_CSTDDEF ${HAVE_CSTDDEF}
+/* HAVE_CSTDINT */
+#cmakedefine HAVE_CSTDINT ${HAVE_CSTDINT}
+/* HAVE_CSTDIO */
+#cmakedefine HAVE_CSTDIO ${HAVE_CSTDIO}
+/* HAVE_CSTDLIB */
+#cmakedefine HAVE_CSTDLIB ${HAVE_CSTDLIB}
+/* HAVE_CASSERT */
+#cmakedefine HAVE_CASSERT ${HAVE_CASSERT}
+/* HAVE_CSTRING */
+#cmakedefine HAVE_CSTRING ${HAVE_CSTRING}
+/* HAVE_CTIME */
+#cmakedefine HAVE_CTIME ${HAVE_CTIME}
+
+/* COIN_C_FINITE */
+#cmakedefine COIN_C_FINITE ${COIN_C_FINITE}
+/* COIN_C_ISNAN */
+#cmakedefine COIN_C_ISNAN ${COIN_C_ISNAN}
+/* COIN_INT64_T */
+#cmakedefine COIN_INT64_T ${COIN_INT64_T}
+/* COIN_UINT64_T */
+#cmakedefine COIN_UINT64_T ${COIN_UINT64_T}
+/* COIN_INTPTR_T */
+#cmakedefine COIN_INTPTR_T ${COIN_INTPTR_T}
+
+/* COIN_${NAME}_CHECKLEVEL */
+#define COIN_${NAME}_CHECKLEVEL ${COIN_${NAME}_CHECKLEVEL}
+/* COIN_${NAME}_VERBOSITY */
+#define COIN_${NAME}_VERBOSITY ${COIN_${NAME}_VERBOSITY}
+
+/* PACKAGE */
+#cmakedefine PACKAGE
+/* PACKAGE_NAME */
+#cmakedefine PACKAGE_NAME
+/* PACKAGE_VERSION */
+#cmakedefine PACKAGE_VERSION
+/* PACKAGE_STRING */
+#cmakedefine PACKAGE_STRING
+/* PACKAGE_TARNAME */
+#cmakedefine PACKAGE_TARNAME
+/* PACKAGE_BUGREPORT */
+#cmakedefine PACKAGE_BUGREPORT
+
diff --git a/Osi/config_osi.h.cmake.in b/Osi/config_osi.h.cmake.in
new file mode 100644
index 0000000..8dc44bf
--- /dev/null
+++ b/Osi/config_osi.h.cmake.in
@@ -0,0 +1,13 @@
+#ifndef __CONFIG_${NAME}_H__
+#define __CONFIG_${NAME}_H__
+
+/* ${NAME}_VERSION */
+#define ${NAME}_VERSION "${VERSION}"
+/* ${NAME}_VERSION_MAJOR */
+#define ${NAME}_VERSION_MAJOR ${MAJOR}
+/* ${NAME}_VERSION_MINOR */
+#define ${NAME}_VERSION_MINOR ${MINOR}
+/* ${NAME}_VERSION_RELEASE */
+#define ${NAME}_VERSION_RELEASE ${PATCH}
+
+#endif
diff --git a/README b/README
index b72bec9..1cd1666 100644
--- a/README
+++ b/README
@@ -1,6 +1,339 @@
 To install this package, please look at the INSTALL file.
- 
+
 This package contains several subdirectories corresponding to COIN-OR
-projects (www.coin-or.org). The AUTHORS, LICENSE and README files in 
+projects (www.coin-or.org). The AUTHORS, LICENSE and README files in
 each of the subdirectories give more information about these projects.
 
+# CBC Version 2.9.8 README
+
+Welcome to the README for the COIN Branch and Cut Solver (CBC). CBC is
+distributed under the Eclipse Public License and is freely redistributable.
+All source code and documentation is Copyright IBM and others. This README may
+be redistributed freely.
+
+## Current Testing Status
+
+[![Build Status](https://travis-ci.org/coin-or/Cbc.svg?branch=master)](https://travis-ci.org/coin-or/Cbc)
+
+[![Build status](https://ci.appveyor.com/api/projects/status/l2hwifsxwhswng8y/branch/master?svg=true)](https://ci.appveyor.com/project/tkralphs/cbc/branch/master)
+
+## DOWNLOAD
+
+[ ![Download](https://api.bintray.com/packages/coin-or/download/Cbc/images/download.svg) ](https://bintray.com/coin-or/download/Cbc/2.9)
+
+Binaries for most platforms are available for download from [Bintray](https://bintray.com/coin-or/download/Cbc)
+
+## DOCUMENTATION
+
+For a quick start guide, please see the INSTALL file in this distribution. A
+(somehwat outdated) user's manual is available here:
+
+http://www.coin-or.org/Cbc
+
+More up-to-date automatically generated documentation of the source code can
+be found here:
+
+http://www.coin-or.org/Doxygen/Cbc/
+
+Further information can be found here:
+
+http://projects.coin-or.org/Cbc
+
+## SUPPORT
+
+### List Serve
+
+CBC users should use the Cbc mailing list. To subscribe, go to
+http://list.coin-or.org/mailman/listinfo/cbc
+
+### Bug Reports
+
+Bug reports should be reported on the CBC development web site at
+
+https://projects.coin-or.org/Cbc/newticket
+
+## CHANGELOG
+
+ * Release 2.9.8
+
+   * Update to most current releases of dependencies
+   * Small bug fixes
+   * Add support for automatic build and test with Travis and Appveyor
+
+ * Release 2.9.7
+
+   * Small bug fixes
+   * Option to switch to line buffered output
+
+ * Release 2.9.6
+
+   * Small bug fixes
+
+ * Release 2.9.5
+
+   * Small bug fixes
+
+ * Release 2.9.4
+
+   * Small fixes for stability
+   * Fixes for Doygen documentation generation
+
+ * Release 2.9.3
+
+   * Minor bug fixes
+
+ * Release 2.9.2
+
+   * Fix for proper installation with ```DESTDIR```
+
+ * Release 2.9.1
+
+   * Fix for dependency linking
+   * Minor bug fixes
+
+ * Release 2.9.0
+
+   * Introduced specialized branching methods for dealing with "big Ms".
+   * Introduced new methods for dealing with symmetry (requires installation of [nauty](http://pallini.di.uniroma1.it/))
+   * Introduction of conflict cuts (off by default, turn on with `-constraint conflict`)
+
+ * Release 2.8.13
+
+   * Improved message handling
+   * Miscellaneous bug fixes.
+
+ * Release 2.8.12
+
+   * Update for dependencies.
+
+ * Release 2.8.11
+
+   * Major overhaul of C interface
+   * Fixes to SOS
+   * Miscellaneous bug fixes
+
+ * Release 2.8.10
+
+   * More changes related to thread safety.
+   * Fix bug in build system with Visual Studio compiler.
+   * Miscellaneous bug fixes.
+
+ * Release 2.8.9
+
+   * Attempt to make Cbc thread safe.
+   * Add parallel examples.
+   * Add CbcSolverUsefulInfo.
+   * Bug fixes.
+
+ * Release 2.8.8
+
+   * Added example to show how to use Cbc with installed libraries in MSVC++
+   * Fixed inconsistency in addition of libCbcSolver to dependencies in
+     {{{cbc_addlibs.txt}}}.
+
+ * Release 2.8.7
+
+   * Changed so that Doxygen builds LaTex
+   * Fixes for build system
+
+ * Release 2.8.6
+
+   * Added option to explicitly link dependencies to comply with packaging
+     requirements on Fedora and Debian, as well as allow building of MinGW
+     DLLs.
+
+ * Release 2.8.5
+
+   * Minor fixes to build system
+
+ * Release 2.8.4
+
+   * Small bug fixes
+   * Upgrades to build system
+
+ * Release 2.8.3:
+
+   * Fix for handling SOS.
+
+ * Release 2.8.2:
+
+   * Fixed recognition of Glpk source in main configure.
+   * Minor bug fixes in CoinUtils, Clp, and Cbc.
+
+ * Release 2.8.1:
+
+   * Minor bug fixes
+
+ * Release 2.8.0:
+
+   * Introduced new secondaryStatus 8 to indicate that solving stopped due to
+     an iteration limit.
+   * Solution pool is now accessible via the command line and the CbcMain*
+     interface.
+   * New mipstart option to read an initial feasible solution from a file.
+     Only values for discrete variables need to be provided.
+
+   * Added Proximity Search heuristic by Fischetti and Monaci (off by
+     default): The simplest way to switch it on using stand-alone version is
+     ```-proximity on```.
+
+     Proximity Search is the new "No-Neighborhood Search" 0-1 MIP refinement
+     heuristic recently proposed by Fischetti and Monaci (2012). The idea is
+     to define a sub-MIP without additional constraints but with a modified
+     objective function intended to attract the search in the proximity of the
+     incumbent. The approach works well for 0-1 MIPs whose solution landscape
+     is not too irregular (meaning the there is reasonable probability of
+     finding an improved solution by flipping a small number of binary
+     variables), in particular when it is applied to the first heuristic
+     solutions found at the root node.
+
+   * An implementation of Zero-Half-Cuts by Alberto Caprara is now available.
+     By default, these cuts are off. To use add to your command line
+     -zerohalfCuts root (or other options) or just -zero. So far, they may
+     help only on a small subset of problems and may need some tuning.
+
+     The implementation of these cuts is described in G. Andreello, A.
+     Caprara, and M. Fischetti "Embedding Cuts in a Branch and Cut Framework:
+     a Computational Study with {0,1/2}-Cuts" INFORMS Journal on Computing
+     19(2), 229-238, 2007 http://dx.doi.org/10.1287/ijoc.1050.0162
+
+   * An alternative implementation of a reduce and split cut generator by
+     Giacomo Nannicini is now available. By default, these cuts are off. To
+     use add to your command line -reduce2AndSplitCuts root (or other
+     options).
+
+     The implementation of these cuts is described in G. Cornuejols and G.
+     Nannicini "Practical strategies for generating rank-1 split cuts in
+     mixed-integer linear programming" Mathematical Programming Computation
+     3(4), 281-318, 2011 http://dx.doi.org/10.1007/s12532-011-0028-6
+
+   * An alternative robust implementation of a Gomory cut generator by Giacomo
+     Nannicini is now available. By default, these cuts are off. To use add to
+     your command line -GMI root (or other options).
+
+     The implementation of these cuts is described in G. Cornuejols, F.
+     Margot, and G. Nannicini "On the safety of Gomory cut generators"
+     http://faculty.sutd.edu.sg/~nannicini/index.php?page=publications
+
+   * To encourage the use of some of the more exotic/expensive cut generators
+     a parameter -slowcutpasses has been added. The idea is that the code does
+     these cuts just a few times - less than the more usual cuts. The default
+     is 10. The cut generators identified by "may be slow" at present are just
+     Lift and project and ReduceAndSplit (both versions).
+
+   * Allow initialization of random seed by user. Pseudo-random numbers are
+     used in Cbc and Clp. In Clp they are used to break ties in degenerate
+     problems, while in Cbc heuristics such as the Feasibility Pump use them
+     to decide whether to round up or down. So if a different pseudo-random
+     seed is given to Clp then you may get a different continuous optimum and
+     so different cuts and heuristic solutions. This can be switched on by
+     setting randomSeed for Clp and/or randomCbcSeed for Cbc. The special
+     value of 0 tells code to use time of day for initial seed.
+
+   * Building on this idea, Andrea Lodi, Matteo Fischetti, Michele Monaci,
+     Domenico Salvagnin, Yuji Shinano, and Andrea Tramontani suggest that this
+     idea be improved by running at the root node with multiple copies of
+     solver, each with its own different seed and then passing in the
+     solutions and cuts so that the main solver has a richer set of solutions
+     and possibly stronger cuts. This is switched on by setting
+     -multipleRootPasses. These can also be done in parallel.
+
+   * Few changes to presolve for special variables and badly scaled problems
+     (in CoinUtils).
+
+   * New option -extraVariables <number> which switches on a trivial
+    re-formulation that introduces extra integer variables to group together
+    variables with same cost.
+
+   * For some problems, cut generators and general branching work better if
+     the problem would be infeasible if the cost is too high. If the new
+     option -constraintFromCutoff is set, the objective function is added as a
+     constraint which rhs is set to the current cutoff value (objective value
+     of best known solution).
+
+ * Release 2.7.8:
+
+   * Change message when LP simplex iteration limit is hit from "Exiting on
+     maximum nodes" to "Exiting on maximum number of iterations"
+   * Fix for using overlapping SOS.
+   * Fixes in buildsystem.
+
+ * Release 2.7.7:
+
+   * Fix to report interruption on user event if SIGINT is received by
+     CbcSolver. model->status() should now be 5 if this event happened. Added
+     method CbcModel::sayEventHappened() to make cbc stop due to an 'user
+     event'.
+
+   * Other minor fixes.
+
+ * Release 2.7.6:
+
+   * Fixes to build system.
+
+   * Other minor fixes.
+
+ * Release 2.7.5:
+
+   * Fixes to get AMPL interface working again.
+
+   * More fixes to MSVC++ files.
+
+ * Release 2.7.4:
+
+   * Minor bugfixes.
+
+ * Release 2.7.3:
+
+   * Minor bugfixes.
+
+   * Fixes to MSVC++ files.
+
+ * Release 2.7.2:
+
+   * Allow row/column names for GMPL models.
+
+   * Added CbcModel::haveMultiThreadSupport() to indicate whether Cbc library
+     has been compiled with multithread support.
+
+   * Added CbcModel::waitingForMiniBranchAndBound() to indicate whether
+     sub-MIP heuristic is currently running.
+
+   * Cbc shell should work with readline if configured with
+     ```--enable-gnu-packages```.
+
+   * Support for compressed input files (.gz, .bz2) is now enabled by default.
+
+   * Fix problems with relative gap tolerance > 100% and further bugs.
+
+   * Fixes for MSVC++ Version 9 files.
+
+   * Minor fixes in buildsystem; update to BuildTools 0.7.1.
+
+ * Release 2.7.1:
+
+   * Fixes to MSVC++ files
+
+ * Release 2.7.0:
+
+   * License has been changed to the EPL.
+
+   * Support for MSVC++ version 10 added.
+
+   * Support for BuildTools version 0.7 to incorporate recent enhancements,
+     including proper library versioning in Linux, prohibiting installation of
+     private headers, etc.
+
+   * Updated externals to new stable versions of dependent projects.
+
+   * Improvements to heuristics.
+
+   * New options for cut generation.
+
+   * Improved reporting of results.
+
+   * Improvements to documentation.
+
+   * Minor bug fixes.
+
+
diff --git a/README.md b/README.md
index c6024f8..981c9b3 100644
--- a/README.md
+++ b/README.md
@@ -1,333 +1,64 @@
-# CBC Version 2.9.8 README
-
-Welcome to the README for the COIN Branch and Cut Solver (CBC). CBC is
-distributed under the Eclipse Public License and is freely redistributable.
-All source code and documentation is Copyright IBM and others. This README may
-be redistributed freely.
-
-## Current Testing Status
-
-[![Build Status](https://travis-ci.org/coin-or/Cbc.svg?branch=master)](https://travis-ci.org/coin-or/Cbc)
-
-[![Build status](https://ci.appveyor.com/api/projects/status/l2hwifsxwhswng8y/branch/master?svg=true)](https://ci.appveyor.com/project/tkralphs/cbc/branch/master)
-
-## DOWNLOAD
-
-[ ![Download](https://api.bintray.com/packages/coin-or/download/Cbc/images/download.svg) ](https://bintray.com/coin-or/download/Cbc/2.9)
-
-Binaries for most platforms are available for download from [Bintray](https://bintray.com/coin-or/download/Cbc)
-
-## DOCUMENTATION
-
-For a quick start guide, please see the INSTALL file in this distribution. A
-(somehwat outdated) user's manual is available here:
-
-http://www.coin-or.org/Cbc
-
-More up-to-date automatically generated documentation of the source code can
-be found here:
-
-http://www.coin-or.org/Doxygen/Cbc/
-
-Further information can be found here:
-
-http://projects.coin-or.org/Cbc
-
-## SUPPORT
-
-### List Serve
-
-CBC users should use the Cbc mailing list. To subscribe, go to 
-http://list.coin-or.org/mailman/listinfo/cbc
-
-### Bug Reports
-
-Bug reports should be reported on the CBC development web site at
-
-https://projects.coin-or.org/Cbc/newticket
-
-## CHANGELOG
-
- * Release 2.9.8
-
-   * Update to most current releases of dependencies
-   * Small bug fixes
-   * Add support for automatic build and test with Travis and Appveyor
-
- * Release 2.9.7
-
-   * Small bug fixes
-   * Option to switch to line buffered output
-
- * Release 2.9.6
-
-   * Small bug fixes
-
- * Release 2.9.5
-
-   * Small bug fixes
-
- * Release 2.9.4
-
-   * Small fixes for stability
-   * Fixes for Doygen documentation generation
-
- * Release 2.9.3
-
-   * Minor bug fixes
-
- * Release 2.9.2
-
-   * Fix for proper installation with ```DESTDIR```
-
- * Release 2.9.1
-
-   * Fix for dependency linking
-   * Minor bug fixes
-
- * Release 2.9.0 
-
-   * Introduced specialized branching methods for dealing with "big Ms".
-   * Introduced new methods for dealing with symmetry (requires installation of [nauty](http://pallini.di.uniroma1.it/))
-   * Introduction of conflict cuts (off by default, turn on with `-constraint conflict`)
-
- * Release 2.8.13
-
-   * Improved message handling
-   * Miscellaneous bug fixes.
-
- * Release 2.8.12
-
-   * Update for dependencies.
-
- * Release 2.8.11
-
-   * Major overhaul of C interface
-   * Fixes to SOS
-   * Miscellaneous bug fixes
-
- * Release 2.8.10
-
-   * More changes related to thread safety.
-   * Fix bug in build system with Visual Studio compiler.
-   * Miscellaneous bug fixes.
-
- * Release 2.8.9
-
-   * Attempt to make Cbc thread safe.
-   * Add parallel examples.
-   * Add CbcSolverUsefulInfo.
-   * Bug fixes.
-
- * Release 2.8.8
-
-   * Added example to show how to use Cbc with installed libraries in MSVC++
-   * Fixed inconsistency in addition of libCbcSolver to dependencies in
-     {{{cbc_addlibs.txt}}}.
-
- * Release 2.8.7
-
-   * Changed so that Doxygen builds LaTex
-   * Fixes for build system
-
- * Release 2.8.6
-
-   * Added option to explicitly link dependencies to comply with packaging
-     requirements on Fedora and Debian, as well as allow building of MinGW
-     DLLs.
-
- * Release 2.8.5
-
-   * Minor fixes to build system
-
- * Release 2.8.4
-
-   * Small bug fixes
-   * Upgrades to build system
-
- * Release 2.8.3:
-
-   * Fix for handling SOS.
-
- * Release 2.8.2:
-
-   * Fixed recognition of Glpk source in main configure.
-   * Minor bug fixes in CoinUtils, Clp, and Cbc.
-
- * Release 2.8.1:
-
-   * Minor bug fixes
-
- * Release 2.8.0:
-
-   * Introduced new secondaryStatus 8 to indicate that solving stopped due to
-     an iteration limit.
-   * Solution pool is now accessible via the command line and the CbcMain*
-     interface.
-   * New mipstart option to read an initial feasible solution from a file.
-     Only values for discrete variables need to be provided.
-
-   * Added Proximity Search heuristic by Fischetti and Monaci (off by
-     default): The simplest way to switch it on using stand-alone version is
-     ```-proximity on```.
-
-     Proximity Search is the new "No-Neighborhood Search" 0-1 MIP refinement
-     heuristic recently proposed by Fischetti and Monaci (2012). The idea is
-     to define a sub-MIP without additional constraints but with a modified
-     objective function intended to attract the search in the proximity of the
-     incumbent. The approach works well for 0-1 MIPs whose solution landscape
-     is not too irregular (meaning the there is reasonable probability of
-     finding an improved solution by flipping a small number of binary
-     variables), in particular when it is applied to the first heuristic
-     solutions found at the root node.
-   
-   * An implementation of Zero-Half-Cuts by Alberto Caprara is now available.
-     By default, these cuts are off. To use add to your command line
-     -zerohalfCuts root (or other options) or just -zero. So far, they may
-     help only on a small subset of problems and may need some tuning.
-
-     The implementation of these cuts is described in G. Andreello, A.
-     Caprara, and M. Fischetti "Embedding Cuts in a Branch and Cut Framework:
-     a Computational Study with {0,1/2}-Cuts" INFORMS Journal on Computing
-     19(2), 229-238, 2007 http://dx.doi.org/10.1287/ijoc.1050.0162
-
-   * An alternative implementation of a reduce and split cut generator by
-     Giacomo Nannicini is now available. By default, these cuts are off. To
-     use add to your command line -reduce2AndSplitCuts root (or other
-     options).
-
-     The implementation of these cuts is described in G. Cornuejols and G.
-     Nannicini "Practical strategies for generating rank-1 split cuts in
-     mixed-integer linear programming" Mathematical Programming Computation
-     3(4), 281-318, 2011 http://dx.doi.org/10.1007/s12532-011-0028-6
-
-   * An alternative robust implementation of a Gomory cut generator by Giacomo
-     Nannicini is now available. By default, these cuts are off. To use add to
-     your command line -GMI root (or other options).
-
-     The implementation of these cuts is described in G. Cornuejols, F.
-     Margot, and G. Nannicini "On the safety of Gomory cut generators"
-     http://faculty.sutd.edu.sg/~nannicini/index.php?page=publications
-
-   * To encourage the use of some of the more exotic/expensive cut generators
-     a parameter -slowcutpasses has been added. The idea is that the code does
-     these cuts just a few times - less than the more usual cuts. The default
-     is 10. The cut generators identified by "may be slow" at present are just
-     Lift and project and ReduceAndSplit (both versions).
-   
-   * Allow initialization of random seed by user. Pseudo-random numbers are
-     used in Cbc and Clp. In Clp they are used to break ties in degenerate
-     problems, while in Cbc heuristics such as the Feasibility Pump use them
-     to decide whether to round up or down. So if a different pseudo-random
-     seed is given to Clp then you may get a different continuous optimum and
-     so different cuts and heuristic solutions. This can be switched on by
-     setting randomSeed for Clp and/or randomCbcSeed for Cbc. The special
-     value of 0 tells code to use time of day for initial seed.
-   
-   * Building on this idea, Andrea Lodi, Matteo Fischetti, Michele Monaci,
-     Domenico Salvagnin, Yuji Shinano, and Andrea Tramontani suggest that this
-     idea be improved by running at the root node with multiple copies of
-     solver, each with its own different seed and then passing in the
-     solutions and cuts so that the main solver has a richer set of solutions
-     and possibly stronger cuts. This is switched on by setting
-     -multipleRootPasses. These can also be done in parallel.
- 
-   * Few changes to presolve for special variables and badly scaled problems
-     (in CoinUtils).
-
-   * New option -extraVariables <number> which switches on a trivial
-    re-formulation that introduces extra integer variables to group together
-    variables with same cost.
-
-   * For some problems, cut generators and general branching work better if
-     the problem would be infeasible if the cost is too high. If the new
-     option -constraintFromCutoff is set, the objective function is added as a
-     constraint which rhs is set to the current cutoff value (objective value
-     of best known solution).
-
- * Release 2.7.8:
-
-   * Change message when LP simplex iteration limit is hit from "Exiting on
-     maximum nodes" to "Exiting on maximum number of iterations"
-   * Fix for using overlapping SOS.
-   * Fixes in buildsystem.
-
- * Release 2.7.7:
-
-   * Fix to report interruption on user event if SIGINT is received by
-     CbcSolver. model->status() should now be 5 if this event happened. Added
-     method CbcModel::sayEventHappened() to make cbc stop due to an 'user
-     event'.
-
-   * Other minor fixes.
-
- * Release 2.7.6:
-
-   * Fixes to build system.
-
-   * Other minor fixes.
-
- * Release 2.7.5:
-
-   * Fixes to get AMPL interface working again.
-
-   * More fixes to MSVC++ files.
-
- * Release 2.7.4:
-
-   * Minor bugfixes.
-
- * Release 2.7.3:
-
-   * Minor bugfixes.
-
-   * Fixes to MSVC++ files.
-
- * Release 2.7.2:
-
-   * Allow row/column names for GMPL models.
-
-   * Added CbcModel::haveMultiThreadSupport() to indicate whether Cbc library
-     has been compiled with multithread support.
-
-   * Added CbcModel::waitingForMiniBranchAndBound() to indicate whether
-     sub-MIP heuristic is currently running.
-
-   * Cbc shell should work with readline if configured with
-     ```--enable-gnu-packages```.
-
-   * Support for compressed input files (.gz, .bz2) is now enabled by default.
-
-   * Fix problems with relative gap tolerance > 100% and further bugs.
-
-   * Fixes for MSVC++ Version 9 files.
-
-   * Minor fixes in buildsystem; update to BuildTools 0.7.1.
-
- * Release 2.7.1:
-
-   * Fixes to MSVC++ files
-
- * Release 2.7.0:
-
-   * License has been changed to the EPL.
-
-   * Support for MSVC++ version 10 added.
-
-   * Support for BuildTools version 0.7 to incorporate recent enhancements,
-     including proper library versioning in Linux, prohibiting installation of
-     private headers, etc.
-
-   * Updated externals to new stable versions of dependent projects.
-
-   * Improvements to heuristics. 
-
-   * New options for cut generation. 
-
-   * Improved reporting of results. 
-
-   * Improvements to documentation. 
-
-   * Minor bug fixes. 
-
+[![Build Status](https://travis-ci.org/Mizux/Cbc.svg?branch=master)](https://travis-ci.org/Mizux/Cbc)
+[![Build status](https://ci.appveyor.com/api/projects/status/rylt2b8cap2ai0ow?svg=true)](https://ci.appveyor.com/project/Mizux/cbc)
+
+# Introduction
+
+This is test of Coin-or Cbc port to [CMake](https://cmake.org/) C++ Project.
+
+This project should run on Linux, Mac and Windows.
+
+# CMake Dependencies Tree
+This CMake project is composed of several targets with the following dependencies:  
+```
+CoinUtils:
+Osi: CoinUils
+Clp: CoinUtils Osi
+OsiClp: Osi Clp
+ClpSolver: Clp
+Cgl: Osi
+Cbc: Cgl Clp
+OsiCbc: Osi Cbc
+CbcSolver: Cbc
+```
+All dependencies are built in static to have one standalone executable.
+## Project directory layout
+Thus the project layout is as follow:
+```
+.
+├── CMakeLists.txt // meta CMake doing the orchestration
+├── cmake
+├── CoinUtils
+│   └── CMakeLists.txt
+├── Osi
+│   └── CMakeLists.txt
+├── Clp
+│   └── CMakeLists.txt
+├── Cgl
+│   └── CMakeLists.txt
+└── Cbc
+    └── CMakeLists.txt
+```
+
+# C++ Project Build
+To build the C++ project, as usual:
+```sh
+cmake -H. -Bbuild
+cmake --build build
+```
+
+# Contributing
+
+The [CONTRIBUTING.md](./CONTRIBUTING.md) file contains instructions on how to
+file the Contributor License Agreement before sending any pull requests (PRs).
+Of course, if you're new to the project, it's usually best to discuss any
+proposals and reach consensus before sending your first PR.
+
+# License
+
+Apache 2 for all patches otherwise see the LICENSE file for details.
+
+# Disclaimer
+
+This is not an official Google product, it is just patches that happens to be
+owned by Google.
 
diff --git a/appveyor.yml b/appveyor.yml
index fb104b0..d91721b 100644
--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,49 +1,27 @@
-platform:
-  - x64
+version: '{build}'
+shallow_clone: true
+
+platform: x64
+configuration: Debug
 
 environment:
-  global:
-    BINTRAY_API:
-      secure: a9n4jf90wlFCdaYa6fOmYxsF97ur2dnK8Ys3gn5R90JBzTDq6cD2GlEwmmts75mq
-    BINTRAY_USERNAME: tkralphs
+  MSBUILD_FLAGS: /verbosity:minimal /maxcpucount
   matrix:
-    - ARCH: win32-msvc9
-      HOST_ARCH_ARG: --enable-msvc=MD
-      ADD_PATH: /mingw64/bin
-    - ARCH: win32-msvc12
-      HOST_ARCH_ARG: --enable-msvc
-      ADD_PATH: /mingw64/bin
-    - ARCH: win32-msvc14
-      HOST_ARCH_ARG: --enable-msvc
-      ADD_PATH: /mingw64/bin
-    - ARCH: x86_64-w64-mingw32
-      HOST_ARCH_ARG: --host=x86_64-w64-mingw32
-      ADD_PATH: /mingw64/bin
-    - ARCH: i686-w64-mingw32
-      HOST_ARCH_ARG: --host=i686-w64-mingw32
-      ADD_PATH: /mingw32/bin
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
+    CMAKE_GENERATOR: "Visual Studio 15 2017 Win64"
+  - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2015
+    CMAKE_GENERATOR: "Visual Studio 14 2015 Win64"
+
+matrix:
+  fast_finish: true
 
-install:
-  - for /f "delims=" %%i in ('C:\msys64\usr\bin\bash -lc "if [ $APPVEYOR_REPO_BRANCH = 'master' ]; then echo 'trunk'; else echo $APPVEYOR_REPO_BRANCH | cut -d "/" -f 2; fi"') do set VERSION=%%i
-  - echo %VERSION%
-  - IF %ARCH%==win32-msvc9 (CALL C:\"Program Files (x86)\Microsoft Visual Studio 9.0\VC\vcvarsall.bat")
-  - IF %ARCH%==win32-msvc12 (CALL C:\"Program Files (x86)\Microsoft Visual Studio 12.0\VC\vcvarsall.bat")
-  - IF %ARCH%==win32-msvc14 (CALL C:\"Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat")
-  - C:\msys64\usr\bin\bash -lc ""
+before_build:
+  - cmake --version
+  - cmake -H. -Bbuild -G "%CMAKE_GENERATOR%"
 
 build_script:
-  - C:\msys64\usr\bin\bash -lc "cd $APPVEYOR_BUILD_FOLDER; git clone https://github.com/coin-or-tools/BuildTools"
-  - C:\msys64\usr\bin\bash -lc "cd $APPVEYOR_BUILD_FOLDER; BuildTools/get.dependencies.sh fetch --no-third-party"
-  - C:\msys64\usr\bin\bash -lc "cd $APPVEYOR_BUILD_FOLDER; export PATH=$ADD_PATH:$PATH; BuildTools/get.dependencies.sh build --build=x86_64-w64-mingw32 $HOST_ARCH_ARG --verbosity=2 --test"
+  - cmake --build build --target ALL_BUILD -- %MSBUILD_FLAGS%
 
-after_build:
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\build\bin
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\build\include
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\build\lib
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\build\share
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\README.md
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\LICENSE
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\INSTALL
-  - 7z a %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip %APPVEYOR_BUILD_FOLDER%\Cbc\AUTHORS
-  - curl -T %APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip -utkralphs:%BINTRAY_API% -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/coin-or/download/%APPVEYOR_PROJECT_NAME%/%VERSION%/%APPVEYOR_PROJECT_NAME%-%VERSION%-%ARCH%.zip
+test_script:
+  - cmake --build build --target RUN_TESTS -- %MSBUILD_FLAGS%
 
diff --git a/cmake/ParseAc.cmake b/cmake/ParseAc.cmake
new file mode 100644
index 0000000..e12dcae
--- /dev/null
+++ b/cmake/ParseAc.cmake
@@ -0,0 +1,11 @@
+function(parse_ac VERSION_STRING VERSION_MAJOR VERSION_MINOR VERSION_PATCH)
+  file(READ "configure.ac" IN)
+  # AC_INIT([Cbc],[2.9.9],[url or email])
+	string(REGEX MATCH
+		"AC_INIT\\([^,]+,\\[([0-9]+)\\.([0-9]+)\\.([0-9]+)\\],[^\\)]+\\)" AC_INIT ${IN})
+	#message(STATUS "AC_INIT: ${AC_INIT}")
+	set(${VERSION_MAJOR} ${CMAKE_MATCH_1} PARENT_SCOPE)
+	set(${VERSION_MINOR} ${CMAKE_MATCH_2} PARENT_SCOPE)
+	set(${VERSION_PATCH} ${CMAKE_MATCH_3} PARENT_SCOPE)
+	set(${VERSION_STRING} "${CMAKE_MATCH_1}.${CMAKE_MATCH_2}.${CMAKE_MATCH_3}" PARENT_SCOPE)
+endfunction()
